{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f8134d3f-60ab-4868-a7dd-a32951790894",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Atomic radii and hydrophobicity values\n",
    "\n",
    "*) radii adapted from parameters use in MSMS\n",
    "*) hydrophobic contribution for amino acid residue atoms and backbone atoms\n",
    "    are based on values obtained by Ghose and Crippen\n",
    "*) values for N- and C-termini not included:\n",
    "    N: -0.4748 (N-term)\n",
    "    C: -0.1703 (C-term)\n",
    "    OT1/2: -0.1962\n",
    "\n",
    "Description v.0.1 (19.01.2010)\n",
    "radii column #0: adjusted vdw radii\n",
    "radii column #1: \"all atom\" Radius set\n",
    "                 based on the \"vdw.hydrogen\" set of radii used in MidasPlus;\n",
    "                 http://www.cgl.ucsf.edu/chimera/1.1700/docs/UsersGuide/midas/vdwtables.html\n",
    "                 assignments are made according to standard atom and residue names.\n",
    "HC  column #2: HC values for HP calculation 1986\n",
    "HC  column #3: HC values for HP calculation 1989\n",
    "HC  column #4: HC values for HP calculation 1998\n",
    "\"\"\"\n",
    "radii = {\n",
    "    (\"ALA\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"ALA\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"ALA\", \" CB \"): [2.17, 1.25, 0.6594, 0.6483, 0.6420, 5],\n",
    "    (\"ALA\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"ALA\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"ARG\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"ARG\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"ARG\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"ARG\", \" CG \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"ARG\", \" CD \"): [2.23, 1.25, -0.0867, 0.0466, 0.2196, 5],\n",
    "    (\"ARG\", \" NE \"): [1.60, 1.45, -0.0886, -0.0455, -0.4204, 1],\n",
    "    (\"ARG\", \" CZ \"): [2.10, 1.50, -0.2692, 0.1847, 0.1388, 0],\n",
    "    (\"ARG\", \" NH1\"): [1.60, 1.45, -0.2817, -0.2894, -0.5851, 1],\n",
    "    (\"ARG\", \" NH2\"): [1.60, 1.45, -0.2817, -0.2894, -0.5851, 1],\n",
    "    (\"ARG\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"ARG\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"ASN\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"ASN\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"ASN\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"ASN\", \" CG \"): [2.10, 1.50, -0.1703, 0.0709, -0.1002, 0],\n",
    "    (\"ASN\", \" OD1\"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"ASN\", \" ND2\"): [1.60, 1.45, -0.5992, -0.7048, -0.7185, 1],\n",
    "    (\"ASN\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"ASN\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"ASP\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"ASP\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"ASP\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"ASP\", \" CG \"): [2.10, 1.50, -0.1703, 0.0709, -0.1002, 0],\n",
    "    (\"ASP\", \" OD1\"): [1.60, 1.35, -0.1962, -0.2587, -0.1216, 2],\n",
    "    (\"ASP\", \" OD2\"): [1.60, 1.35, -0.1962, -0.2587, -0.1216, 2],\n",
    "    (\"ASP\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"ASP\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"CYS\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"CYS\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"CYS\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"CYS\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"CYS\", \" CB \"): [2.23, 1.25, -0.0867, 0.0466, 0.2196, 5],\n",
    "    (\"CYS\", \" SG \"): [1.89, 1.35, 0.6449, 0.4008, 0.5110, 5],\n",
    "    (\"GLU\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"GLU\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"GLU\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"GLU\", \" CG \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"GLU\", \" CD \"): [2.10, 1.50, -0.1703, 0.0709, -0.1002, 0],\n",
    "    (\"GLU\", \" OE1\"): [1.60, 1.35, -0.1962, -0.2587, -0.1216, 2],\n",
    "    (\"GLU\", \" OE2\"): [1.60, 1.35, -0.1962, -0.2587, -0.1216, 2],\n",
    "    (\"GLU\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"GLU\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"GLN\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"GLN\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"GLN\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"GLN\", \" CG \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"GLN\", \" CD \"): [2.10, 1.50, -0.1703, 0.0709, -0.1002, 0],\n",
    "    (\"GLN\", \" OE1\"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"GLN\", \" NE2\"): [1.60, 1.45, -0.5992, -0.7048, -0.7185, 1],\n",
    "    (\"GLN\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"GLN\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"GLY\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"GLY\", \" CA \"): [2.23, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"GLY\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"GLY\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"HIS\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"HIS\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"HIS\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"HIS\", \" CG \"): [2.10, 1.70, 0.3345, 0.1600, 0.1492, 4],\n",
    "    (\"HIS\", \" CD2\"): [2.10, 1.70, 0.3952, 0.1569, 0.2578, 4],\n",
    "    (\"HIS\", \" ND1\"): [1.60, 1.70, -0.0210, 0.0938, 0.0223, 3],\n",
    "    (\"HIS\", \" CE1\"): [2.10, 1.70, -0.3509, 0.2027, 0.4154, 4],\n",
    "    (\"HIS\", \" NE2\"): [1.60, 1.70, 0.3493, 0.4198, 0.1259, 7],\n",
    "    (\"HIS\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"HIS\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"ILE\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"ILE\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"ILE\", \" CB \"): [2.37, 1.25, 0.1514, 0.0785, 0.0660, 5],\n",
    "    (\"ILE\", \" CG2\"): [2.17, 1.25, 0.6594, 0.6483, 0.6420, 5],\n",
    "    (\"ILE\", \" CG1\"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"ILE\", \" CD1\"): [2.17, 1.25, 0.6594, 0.6483, 0.6420, 5],\n",
    "    (\"ILE\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"ILE\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"LEU\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"LEU\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"LEU\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"LEU\", \" CG \"): [2.37, 1.25, 0.1514, 0.0785, 0.0660, 5],\n",
    "    (\"LEU\", \" CD1\"): [2.17, 1.70, 0.6594, 0.6483, 0.6420, 5],\n",
    "    (\"LEU\", \" CD2\"): [2.17, 1.70, 0.6594, 0.6483, 0.6420, 5],\n",
    "    (\"LEU\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"LEU\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"LYS\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"LYS\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"LYS\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"LYS\", \" CG \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"LYS\", \" CD \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"LYS\", \" CE \"): [2.23, 1.25, -0.0867, 0.0466, 0.2196, 5],\n",
    "    (\"LYS\", \" NZ \"): [1.60, 1.25, -0.4748, -0.5333, -0.7499, 1],\n",
    "    (\"LYS\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"LYS\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"MET\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"MET\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"MET\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"MET\", \" CG \"): [2.23, 1.25, -0.0867, 0.0466, 0.2196, 5],\n",
    "    (\"MET\", \" SD \"): [1.89, 1.35, 1.0339, 0.6145, 0.5906, 5],\n",
    "    (\"MET\", \" CE \"): [2.17, 1.25, 0.1460, 0.2430, 0.4143, 5],\n",
    "    (\"MET\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"MET\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"PHE\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"PHE\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"PHE\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"PHE\", \" CG \"): [2.10, 1.70, 0.3345, 0.1600, 0.1492, 4],\n",
    "    (\"PHE\", \" CD1\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"PHE\", \" CD2\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"PHE\", \" CE1\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"PHE\", \" CE2\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"PHE\", \" CZ \"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"PHE\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"PHE\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"PRO\", \" N  \"): [1.80, 1.45, 0.3990, 0.3954, 0.0132, 0],\n",
    "    (\"PRO\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"PRO\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"PRO\", \" CG \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"PRO\", \" CD \"): [2.23, 1.25, -0.0867, 0.0466, 0.2196, 5],\n",
    "    (\"PRO\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"PRO\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"SER\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"SER\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"SER\", \" CB \"): [2.23, 1.25, -0.0867, 0.0466, 0.2196, 0],\n",
    "    (\"SER\", \" OG \"): [1.60, 1.35, -0.4220, -0.1858, -0.4603, 3],\n",
    "    (\"SER\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"SER\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"THR\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"THR\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"THR\", \" CB \"): [2.37, 1.25, -0.5156, -0.0792, 0.0536, 0],\n",
    "    (\"THR\", \" OG1\"): [1.60, 1.35, -0.4220, -0.1858, -0.4603, 3],\n",
    "    (\"THR\", \" CG2\"): [2.17, 1.25, 0.6594, 0.6483, 0.6420, 5],\n",
    "    (\"THR\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"THR\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"TRP\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"TRP\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"TRP\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 0],\n",
    "    (\"TRP\", \" CG \"): [2.10, 1.70, 0.3345, 0.1600, 0.1492, 4],\n",
    "    (\"TRP\", \" CD2\"): [2.10, 1.70, 0.3345, 0.1600, 0.1492, 4],\n",
    "    (\"TRP\", \" CE2\"): [2.10, 1.70, 0.2455, -0.2782, 0.2813, 4],\n",
    "    (\"TRP\", \" CE3\"): [2.10, 1.70, 0.3147, 0.3411, 0.3050, 4],\n",
    "    (\"TRP\", \" CD1\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"TRP\", \" NE1\"): [1.60, 1.70, -0.1946, -0.4366, -0.2660, 6],\n",
    "    (\"TRP\", \" CZ2\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"TRP\", \" CZ3\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"TRP\", \" CH2\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"TRP\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"TRP\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"TYR\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"TYR\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"TYR\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 4],\n",
    "    (\"TYR\", \" CG \"): [2.10, 1.70, 0.3345, 0.1600, 0.1492, 0],\n",
    "    (\"TYR\", \" CD1\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"TYR\", \" CE1\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"TYR\", \" CD2\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"TYR\", \" CE2\"): [2.10, 1.70, 0.3174, 0.3411, 0.3050, 4],\n",
    "    (\"TYR\", \" CZ \"): [2.10, 1.70, -0.1153, -0.1033, 0.1539, 4],\n",
    "    (\"TYR\", \" OH \"): [1.60, 1.35, 0.1509, 0.1600, -0.1163, 3],\n",
    "    (\"TYR\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"TYR\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"VAL\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],\n",
    "    (\"VAL\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"VAL\", \" CB \"): [2.37, 1.25, 0.1514, 0.0785, 0.0660, 5],\n",
    "    (\"VAL\", \" CG1\"): [2.17, 1.25, 0.6594, 0.6483, 0.6420, 5],\n",
    "    (\"VAL\", \" CG2\"): [2.17, 1.25, 0.6594, 0.6483, 0.6420, 5],\n",
    "    (\"VAL\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"VAL\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (\"MSE\", \" N  \"): [1.80, 1.45, -0.2289, -0.3788, -0.6149, 1],  # selenomethionine\n",
    "    (\"MSE\", \" CA \"): [2.37, 1.25, -0.4621, -0.1012, 0.3663, 0],\n",
    "    (\"MSE\", \" CB \"): [2.23, 1.25, 0.4616, 0.3963, 0.4562, 5],\n",
    "    (\"MSE\", \" CG \"): [2.23, 1.25, -0.0867, 0.0466, 0.2196, 5],\n",
    "    (\"MSE\", \" CE \"): [2.17, 1.25, 0.1460, 0.2430, 0.4143, 5],\n",
    "    (\"MSE\", \"SE  \"): [1.89, 1.35, 1.0339, 0.6145, 0.5906, 5],\n",
    "    (\"MSE\", \" C  \"): [1.70, 1.50, -0.1703, 0.0709, -0.1002, 4],\n",
    "    (\"MSE\", \" O  \"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 2],\n",
    "    (None, \" OXT\"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 0],\n",
    "    (None, \"XE  \"): [2.16, 2.16, 0.0, 0.0, 0.0, 0],  # Xenon\n",
    "    (None, \"CA  \"): [2.00, 2.00, 0.0, 0.0, 0.0, 0],  # Calcium\n",
    "    (None, \"ZN  \"): [2.00, 2.00, 0.0, 0.0, 0.0, 0],  # Zinc\n",
    "    (None, \" C2\"): [1.70, 1.70, 0.0, 0.0, 0.0, 0],\n",
    "    (None, \" C4\"): [1.70, 1.70, 0.0, 0.0, 0.0, 0],\n",
    "    (None, \" C5\"): [1.70, 1.70, 0.0, 0.0, 0.0, 0],\n",
    "    (None, \" C6\"): [1.70, 1.70, 0.0, 0.0, 0.0, 0],\n",
    "    (None, \" C8\"): [1.70, 1.70, 0.0, 0.0, 0.0, 0],\n",
    "    (None, \" C \"): [1.70, 1.25, 0.0, 0.0, 0.0, 0],\n",
    "    (None, \" C\"): [1.70, 1.25, 0.0, 0.0, 0.0, 0],  # all C's if C2-C8 not found\n",
    "    (None, \" N4\"): [1.60, 1.45, 0.0, 0.0, 0.0, 0],\n",
    "    (None, \" N\"): [1.60, 1.70, 0.0, 0.0, 0.0, 0],  # all N's if NX not found\n",
    "    (None, \" O\"): [1.60, 1.35, 0.0, 0.0, 0.0, 0],  # all remaining O's\n",
    "    (None, \" P\"): [2.00, 1.70, 0.0, 0.0, 0.0, 0],  # all remaining P's\n",
    "    (None, \" S\"): [2.00, 1.70, 0.0, 0.0, 0.0, 0],  # all remaining S's\n",
    "    (None, \" OT1\"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 0],  # Terminal Oxygen 1\n",
    "    (None, \" OT2\"): [1.60, 1.35, -0.1729, -0.3514, -0.0233, 0],  # Terminal Oxygen 2\n",
    "    (None, \" H\"): [1.00, 1.00, 0.0, 0.0, 0.0, 0],  # Hydrogen\n",
    "}\n",
    "\n",
    "charges = {\n",
    "    (\"ARG\", \" NE \"): 1.0 / 3.0,\n",
    "    (\"ARG\", \" NH1\"): 1.0 / 3.0,\n",
    "    (\"ARG\", \" NH2\"): 1.0 / 3.0,\n",
    "    (\"ASP\", \" OD1\"): -0.5,\n",
    "    (\"ASP\", \" OD2\"): -0.5,\n",
    "    (\"GLU\", \" OE1\"): -0.5,\n",
    "    (\"GLU\", \" OE2\"): -0.5,\n",
    "    (\"HIS\", \" ND1\"): 1.0 / 6.0,\n",
    "    (\"HIS\", \" NE2\"): 1.0 / 6.0,\n",
    "    (\"LYS\", \" NZ \"): 1.0,\n",
    "    # (None, \" N  \"): -0.084,  # peptide bond\n",
    "    # (None, \" C  \"): 0.506,\n",
    "    # (None, \" O  \"): -0.422,\n",
    "    (None, \" OXT\"): -1.0,  # C-terminal carboxylate\n",
    "    (None, \" OT1\"): -0.5,\n",
    "    (None, \" OT2\"): -0.5,\n",
    "    (None, \"MG  \"): 2.0,  # Magnesium\n",
    "    (None, \"CA  \"): 2.0,  # Calcium\n",
    "    (None, \"ZN  \"): 2.0,  # Zinc\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6004fd6d-ee3a-43f7-a891-200abf8e2974",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "class definitions for molecular structures (in PDB format)\n",
    "\"\"\"\n",
    "from io import TextIOWrapper  # necessary to check for file objects\n",
    "from logging import getLogger\n",
    "\n",
    "\n",
    "logger = getLogger(__name__)\n",
    "\n",
    "\n",
    "class PDBAtom:\n",
    "    \"\"\"\n",
    "    Class for a single atom in a (protein) structure.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, line=None):\n",
    "        # extracts information from a line of PDB formatted text,\n",
    "        # no error checking is performed.\n",
    "        self.het = False\n",
    "        self.atom_number = 0\n",
    "        self.name = \" X \"\n",
    "        self.alternate = \"\"\n",
    "        self.residue = \"UNK\"\n",
    "        self.chain = \"X\"\n",
    "        self.residue_number = 0\n",
    "        self.insertion = \" \"\n",
    "        self.x = 0.0\n",
    "        self.y = 0.0\n",
    "        self.z = 0.0\n",
    "        self.occupancy = 0.0\n",
    "        self.bfactor = 0.0\n",
    "\n",
    "        self.segi = \" \"\n",
    "        self.element = \" \"\n",
    "        self.crg = \"  \"\n",
    "\n",
    "        # other possible atom parameters\n",
    "        self.keep_old_bfactor = self.bfactor\n",
    "        self.HP = 0.0\n",
    "        self.radius = 2.0\n",
    "        self.charge = 0.0\n",
    "        self.formal_charge = 0.0\n",
    "        self.par1 = 0.0\n",
    "        self.par2 = 0.0\n",
    "        self.pseudo_id = 0\n",
    "\n",
    "        # property dictionary: {\"prop1\": value,\"prop2\": value\"}\n",
    "        self.prop_dic = {}\n",
    "\n",
    "        if line is not None:  # read data from a PDB ATOM or HETATM line\n",
    "            self.add_line(line)\n",
    "\n",
    "    def add_line(self, line):\n",
    "        \"\"\"\n",
    "        Read atom data from a line of PDB-formatted text,\n",
    "        assumes line to start with 'ATOM' or 'HETATM'\n",
    "        \"\"\"\n",
    "        self.het = line[:6] == \"HETATM\"\n",
    "        self.atom_number = int(line[6:11])\n",
    "        self.name = line[12:16]\n",
    "        self.alternate = line[16]\n",
    "        self.residue = line[17:20]\n",
    "        self.chain = line[21]\n",
    "        self.residue_number = line[22:26]\n",
    "        self.insertion = line[26]\n",
    "        self.x = float(line[30:38])\n",
    "        self.y = float(line[38:46])\n",
    "        self.z = float(line[46:54])\n",
    "\n",
    "        # in some PDB files, everything beyond the coordinates is missing\n",
    "        try:\n",
    "            self.occupancy = float(line[54:60])\n",
    "        except ValueError:\n",
    "            self.occupancy = 1.0\n",
    "\n",
    "        try:\n",
    "            self.bfactor = float(line[60:66])\n",
    "        except ValueError:\n",
    "            self.bfactor = 20.0\n",
    "\n",
    "        self.segi = line[71:76]\n",
    "        self.element = line[76:78]\n",
    "        self.crg = line[78:80]\n",
    "        self.keep_old_bfactor = self.bfactor\n",
    "\n",
    "    def set_prop(self, prop, value=None):\n",
    "        \"\"\"\n",
    "        Set property 'prop' of atom to value.\n",
    "        \"\"\"\n",
    "        if hasattr(self, prop):\n",
    "            setattr(self, prop, value)\n",
    "\n",
    "    def add_prop_dic(self, prop, value=None):\n",
    "        \"\"\"\n",
    "        Add or update custom property dictionary.\n",
    "        \"\"\"\n",
    "        self.prop_dic.update({prop: value})\n",
    "\n",
    "    def _print_prop_info(self):\n",
    "        u = []\n",
    "        u.extend(\n",
    "            [\n",
    "                \"keep_old_bfactor\",\n",
    "                \"radius          \",\n",
    "                \"charge          \",\n",
    "                \"formal_charge   \",\n",
    "                \"par1            \",\n",
    "                \"par2            \",\n",
    "                \"HP              \",\n",
    "                \"pseudo_id       \",\n",
    "            ]\n",
    "        )\n",
    "        u.extend([key for key in self.prop_dic.keys()])\n",
    "        u = [\"%-16s\" % (str(i)) for i in u]\n",
    "        print(\" \".join(u))\n",
    "\n",
    "    def print_prop(self, form=\"table\"):\n",
    "        \"\"\"\n",
    "        Output non-standard properties in various formats\n",
    "        form: 'table'\n",
    "              'catalophore'\n",
    "              'csv'\n",
    "        \"\"\"\n",
    "        if form == \"table\":\n",
    "            print(self.__str__())\n",
    "            print(\"Properties:\")\n",
    "            print(\"-----------\")\n",
    "            print(\".keep_old_bfactor:   %s\" % self.keep_old_bfactor)\n",
    "            print(\".radius          :   %s\" % self.radius)\n",
    "            print(\".charge          :   %s\" % self.charge)\n",
    "            print(\".formal_charge   :   %s\" % self.formal_charge)\n",
    "            print(\".par1            :   %s\" % self.par1)\n",
    "            print(\".par2            :   %s\" % self.par2)\n",
    "            print(\".HP              :   %s\" % self.HP)\n",
    "            print(\".pseudo_id       :   %s\" % self.pseudo_id)\n",
    "            print(\"-------------------------\")\n",
    "            print(\"Additional: .prop_dic\")\n",
    "            for key, value in self.prop_dic.items():\n",
    "                print(\"%s: %s\" % (key, value))\n",
    "\n",
    "        elif form == \"catalophore\":\n",
    "            print(\n",
    "                \",\".join(\n",
    "                    [\n",
    "                        str(a)\n",
    "                        for a in [\n",
    "                            self.atom_number,\n",
    "                            self.residue_number,\n",
    "                            self.x,\n",
    "                            self.y,\n",
    "                            self.z,\n",
    "                            self.keep_old_bfactor,\n",
    "                            self.HP,\n",
    "                        ]\n",
    "                    ]\n",
    "                )\n",
    "            )\n",
    "\n",
    "        else:\n",
    "            u = [\n",
    "                int(self.atom_number),  # point_id\n",
    "                int(self.residue_number),  # cav_id\n",
    "                float(self.x),  # x\n",
    "                float(self.y),  # y\n",
    "                float(self.z),  # z\n",
    "            ]\n",
    "            u.extend(\n",
    "                [\n",
    "                    self.keep_old_bfactor,\n",
    "                    self.radius,\n",
    "                    self.charge,\n",
    "                    self.formal_charge,\n",
    "                    self.par1,\n",
    "                    self.par2,\n",
    "                    self.HP,\n",
    "                    self.pseudo_id,\n",
    "                ]\n",
    "            )\n",
    "            u.extend([value for value in self.prop_dic.values()])\n",
    "            u = [\"%s\" % (str(i)) for i in u]\n",
    "            print(\",\".join(u))\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.get_pdbstr()\n",
    "\n",
    "    def get_pdbstr(self, prop=None, prop_key=None):\n",
    "        \"\"\"\n",
    "        Return a PDB-formatted line.\n",
    "        The B-factor column can be replaced by the value in 'prop' or an entry from 'prop_dic'\n",
    "        (if prop == 'prop_dic')\n",
    "\n",
    "        If property of dictionary key is not found, the B-factor column is set to -1.\n",
    "        :param prop: property of PDBAtom object or 'prop_dic'\n",
    "        :param prop_key: key in PDBAtom.prop_dic\n",
    "        :return: PDB-formatted line\n",
    "        \"\"\"\n",
    "        if self.het:\n",
    "            start = \"HETATM\"\n",
    "        else:\n",
    "            start = \"ATOM  \"\n",
    "\n",
    "        if prop is None or prop == \"bfactor\":  # use the value im atom.bfactor\n",
    "            value = self.bfactor\n",
    "        elif prop == \"prop_dic\":  # use a value from 'prop_dic'\n",
    "            value = self.prop_dic.get(prop_key, -1.0)\n",
    "        else:  # use another property\n",
    "            if hasattr(self, prop):\n",
    "                value = getattr(self, prop)\n",
    "            else:\n",
    "                value = -1.0\n",
    "\n",
    "        return (\n",
    "            \"%-6s%5d %4s%1s%3s %1s%4d%1s   %8.3f%8.3f%8.3f%6.2f%6.2f      %-4s%2s%2s\"\n",
    "            % (\n",
    "                start,\n",
    "                int(self.atom_number),\n",
    "                str(self.name),\n",
    "                str(self.alternate),\n",
    "                str(self.residue),\n",
    "                str(self.chain),\n",
    "                int(self.residue_number),\n",
    "                str(self.insertion),\n",
    "                float(self.x),\n",
    "                float(self.y),\n",
    "                float(self.z),\n",
    "                float(self.occupancy),\n",
    "                value,  # float(self.bfactor),  replacement value for the B-factor\n",
    "                str(self.segi),\n",
    "                str(self.element),\n",
    "                str(self.crg),\n",
    "            )\n",
    "        )\n",
    "\n",
    "\n",
    "class PDBStructure:\n",
    "    \"\"\"\n",
    "    Class for a protein structure which is read from a file-like,\n",
    "    PDB-formatted object.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, file_obj=None):\n",
    "        \"\"\"\n",
    "        reads a PDB file line per line. Non-atom lines are stored in a\n",
    "        separate list.\n",
    "\n",
    "        file_obj can be a file object or a list (tuple) of PDB lines\n",
    "        \"\"\"\n",
    "        self.remarks = []  # list that contains non-atom lines\n",
    "        self.atom = []  # list of instances of PdbAtom\n",
    "        self.water = []  # water coordinates\n",
    "\n",
    "        # The following attributes are currently not needed\n",
    "        # self.deleted_atom = []      # as atom[], but for deleted atoms (H, HETATM,...)\n",
    "        # self.dummy = []             # dummy coordinates\n",
    "\n",
    "        # # pseudo center atoms (CavBase-like pseudo atoms)\n",
    "        # self.pseudo_aliphatic = []  # aliphatic pseudo atoms\n",
    "        # self.pseudo_pi = []         # pi pseudo atoms\n",
    "        # self.pseudo_donor = []      # donor atoms\n",
    "        # self.pseudo_acceptor = []   # acceptor atoms\n",
    "        # self.pseudo_DON_ACC = []    # donor_acceptor atoms\n",
    "\n",
    "        if file_obj is not None:\n",
    "            if (\n",
    "                isinstance(file_obj, list)\n",
    "                or isinstance(file_obj, tuple)\n",
    "                or isinstance(file_obj, TextIOWrapper)\n",
    "            ):  # 'file'\n",
    "                tmp = file_obj\n",
    "            elif isinstance(file_obj, str):\n",
    "                tmp = file_obj.split(\"\\n\")  # assuming PyMOL pdbstr\n",
    "            else:\n",
    "                return\n",
    "\n",
    "            for line in tmp:\n",
    "                if self._is_atom(line):\n",
    "                    self.atom.append(PDBAtom(line))\n",
    "                else:\n",
    "                    self.remarks.append(line)\n",
    "\n",
    "            self.add_water()\n",
    "\n",
    "    @staticmethod\n",
    "    def _is_atom(line):\n",
    "        \"\"\"Checks whether the line contains atom data.\"\"\"\n",
    "        return line[:4] == \"ATOM\" or line[:6] == \"HETATM\"\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.get_pdbstr()\n",
    "\n",
    "    def get_pdbstr(self, prop=None, prop_key=None):\n",
    "        \"\"\"\n",
    "        Return pdbstr of atoms\n",
    "        \"\"\"\n",
    "        return (\n",
    "            \"\\n\".join([i.get_pdbstr(prop=prop, prop_key=prop_key) for i in self.atom])\n",
    "            + \"\\n\"\n",
    "        )\n",
    "\n",
    "    def add_water(self, atoms=None, resn_list=(\"H2O\", \"HOH\", \"WAT\")):\n",
    "        \"\"\"\n",
    "        Put water atoms to separate atoms list called water.\n",
    "        atoms: list of atom objects\n",
    "        resn_list: (\"H2O\",\"HOH\",\"WAT\") residue names for water molecules.\n",
    "        \"\"\"\n",
    "        if atoms is None:\n",
    "            atoms = self.atom\n",
    "\n",
    "        for atom in atoms:\n",
    "            if atom.residue in resn_list:\n",
    "                self.water.append(atom)\n",
    "\n",
    "    def change_bfac(\n",
    "        self, atom_list=None, from_prop=\"par1\", prop_key=None, to_prop=\"bfactor\"\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Change the bfac field from old property to new property.\n",
    "        atomlist: list of PdbAtom objects (standard: self.atom)\n",
    "\n",
    "        from_prop: \"par1\",\n",
    "                   \"par2\",\n",
    "                   \"radius\"\n",
    "                   \"charge\"\n",
    "                   \"occupancy\"\n",
    "                   ...         other properties have to be added\n",
    "                   \"prop_dic\" additional property dictionary accessible by\n",
    "                              \"prop_key\" value\n",
    "\n",
    "\n",
    "                   \"restore\"   restore old b-factor value\n",
    "        to_prop:   \"bfactor\"   currently only b-factor can be switched\n",
    "                               old b-factor is still in \"keep_old_bfactor\"\n",
    "                               (KG: This is obviously not true, anymore. See code!)\n",
    "        \"\"\"\n",
    "        if atom_list is None:\n",
    "            atom_list = self.atom\n",
    "        elif hasattr(\n",
    "            self, atom_list\n",
    "        ):  # in case the PDBStructure object has an attribute 'atom_list'\n",
    "            atom_list = getattr(self, atom_list)\n",
    "\n",
    "        if len(atom_list) == 0:\n",
    "            logger.warning(\"No atoms in selection.\")\n",
    "            return\n",
    "\n",
    "        if from_prop == \"prop_dic\":  # read from property dictionary\n",
    "            if prop_key:\n",
    "                if atom_list[0].prop_dic.get(prop_key) is not None:\n",
    "                    if hasattr(atom_list[0], to_prop):  # check atom[0]\n",
    "                        for atom in atom_list:\n",
    "                            setattr(atom, to_prop, atom.prop_dic.get(prop_key))\n",
    "                    else:\n",
    "                        logger.warning(\"No %s property found in atom.\", str(to_prop))\n",
    "                else:\n",
    "                    logger.warning(\n",
    "                        \"No %s property found in atom prop_dic.\", str(prop_key)\n",
    "                    )\n",
    "                    return\n",
    "\n",
    "        elif from_prop == \"restore\":  # restore original B-factor\n",
    "            for atom in atom_list:\n",
    "                atom.bfactor = atom.keep_old_bfactor\n",
    "\n",
    "        elif hasattr(atom_list[0], from_prop) and hasattr(  # check atom[0]\n",
    "            atom_list[0], to_prop\n",
    "        ):\n",
    "            for atom in atom_list:\n",
    "                setattr(atom, to_prop, getattr(atom, from_prop))\n",
    "\n",
    "    def assign_rad_hp_chg(\n",
    "        self, HP_dict=None, HP_column=2, rad_dict=None, rad_column=0, chg_dict=None\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Assign radii and hydrophobicity values to atoms in a structure\n",
    "        :param chg_dict:\n",
    "        :param HP_dict:\n",
    "        :param HP_column:\n",
    "        :param rad_dict:\n",
    "        :param rad_column:\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        if rad_dict is None:\n",
    "            rad_dict = radii  # internal radius dictionary\n",
    "        if HP_dict is None:\n",
    "            HP_dict = radii  # internal radius/HP dictionary\n",
    "        if chg_dict is None:\n",
    "            chg_dict = charges\n",
    "\n",
    "        for atom in self.atom:\n",
    "            queries = [  # atom queries with decreasing specificity\n",
    "                (atom.residue, atom.name),\n",
    "                (None, atom.name),\n",
    "                (None, atom.name[:3]),\n",
    "                (None, atom.name[:2]),\n",
    "            ]\n",
    "\n",
    "            # set atom radii\n",
    "            for query in queries:\n",
    "                value = rad_dict.get(query)\n",
    "                if value:\n",
    "                    atom.radius = value[rad_column]\n",
    "                    break\n",
    "            else:\n",
    "                logger.warning(\n",
    "                    \"ATOM %s %s %s unknown. Radius is set to %.1f\"\n",
    "                    % (atom.atom_number, atom.residue, atom.name, 2.0)\n",
    "                )\n",
    "                atom.radius = 2.0\n",
    "\n",
    "            # set HC values\n",
    "            for query in queries:\n",
    "                value = HP_dict.get(query)\n",
    "                if value:\n",
    "                    atom.HP = value[HP_column]\n",
    "                    break\n",
    "            else:\n",
    "                logger.warning(\n",
    "                    \"ATOM %s %s %s unknown. HC value is set to %.1f\"\n",
    "                    % (atom.atom_number, atom.residue, atom.name, 0.0)\n",
    "                )\n",
    "                atom.HP = 0.0\n",
    "\n",
    "            # set partial charges\n",
    "            for query in queries:\n",
    "                value = chg_dict.get(query)\n",
    "                if value:\n",
    "                    atom.charge = value\n",
    "                    break\n",
    "            else:\n",
    "                atom.charge = 0.0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e1cae2af-456d-415a-949f-8793efda53ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "function definitions for LigSite and cavity detection\n",
    "\"\"\"\n",
    "import itertools\n",
    "import math\n",
    "\n",
    "import numpy as np\n",
    "from numpy.lib.stride_tricks import as_strided\n",
    "\n",
    "\n",
    "\n",
    "# 32-bit precision provides speed-up in HP annotation\n",
    "FP_DTYPE = np.float32\n",
    "\n",
    "\n",
    "def setup_grid(coords, cushion, d_grid, init, dtype):\n",
    "    \"\"\"\n",
    "    Setup and initialize grid using atom coordinates\n",
    "    \"\"\"\n",
    "    # min and max in cartesian coordinates\n",
    "    min_coords = coords.min(axis=0)\n",
    "    max_coords = coords.max(axis=0)\n",
    "    # min and max in grid coordinates\n",
    "    min_grid = np.floor((min_coords - cushion) / d_grid)\n",
    "    max_grid = np.ceil((max_coords + cushion) / d_grid)\n",
    "\n",
    "    origin = min_grid * d_grid\n",
    "    extent = (max_grid - min_grid + 1).astype(int)\n",
    "\n",
    "    return Grid(origin=origin, extent=extent, d=d_grid, init=init, dtype=dtype)\n",
    "\n",
    "\n",
    "def mask_grid(\n",
    "    grid, coords, radii, radius_factor, probe_radius, softness, protein_flag, soft_flag\n",
    "):\n",
    "    \"\"\"\n",
    "    Mask grid using atom coordinates and radii\n",
    "    \"\"\"\n",
    "    # atom coordinates in grid units\n",
    "    grid_coords = (coords - grid.origin) / grid.d\n",
    "    # atom radii in grid units\n",
    "    grid_radii = (radii * radius_factor + probe_radius) / grid.d\n",
    "    r1sq = grid_radii**2  # squared outer, soft radius\n",
    "    r2sq = (grid_radii - softness) ** 2  # squared inner, hard radius\n",
    "\n",
    "    # origin and space-diagonal coordinates of the sub-grids around the atoms\n",
    "    sg_start = np.clip(\n",
    "        np.floor(grid_coords - grid_radii.reshape(-1, 1)).astype(int),\n",
    "        (0, 0, 0),\n",
    "        grid.extent,\n",
    "    )\n",
    "    sg_end = np.clip(\n",
    "        np.ceil(grid_coords + grid_radii.reshape(-1, 1)).astype(int) + 1,\n",
    "        (0, 0, 0),\n",
    "        grid.extent,\n",
    "    )\n",
    "\n",
    "    for i in range(grid_coords.shape[0]):  # loop over all atom coordinates\n",
    "        x_start, y_start, z_start = sg_start[i]\n",
    "        x_end, y_end, z_end = sg_end[i]\n",
    "        sub_grid = grid.get_subgrid(x_start, y_start, z_start, x_end, y_end, z_end)\n",
    "\n",
    "        x, y, z = np.ogrid[x_start:x_end, y_start:y_end, z_start:z_end]\n",
    "        dist = (\n",
    "            (x - grid_coords[i, 0]) ** 2\n",
    "            + (y - grid_coords[i, 1]) ** 2\n",
    "            + (z - grid_coords[i, 2]) ** 2\n",
    "        )\n",
    "\n",
    "        # grid points within the inner, hard radius\n",
    "        mask_hard = dist < r2sq[i]\n",
    "        sub_grid[mask_hard] = protein_flag\n",
    "\n",
    "        # grid points between inner and outer radius\n",
    "        mask_soft = (dist < r1sq[i]) & (sub_grid != protein_flag)\n",
    "        sub_grid[mask_soft] = soft_flag\n",
    "\n",
    "\n",
    "def do_ligsite(masked_grid, protein_flag):\n",
    "    \"\"\"\n",
    "    Run ligsite algorithm.\n",
    "    Calculate LigSite scores for a masked grid\n",
    "    \"\"\"\n",
    "    grid = masked_grid.get_grid()\n",
    "    nx = masked_grid.nx\n",
    "    ny = masked_grid.ny\n",
    "    nz = masked_grid.nz\n",
    "    offset = masked_grid.offset\n",
    "\n",
    "    # analyze along the x-direction\n",
    "    for iy in range(ny):\n",
    "        for iz in range(nz):\n",
    "            line = grid[:, iy, iz]\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # analyze along the y-direction\n",
    "    for ix in range(nx):\n",
    "        for iz in range(nz):\n",
    "            line = grid[ix, :, iz]\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # analyze along the z-direction\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            line = grid[ix, iy, :]\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # space-diagonal 1 (1,1,1)\n",
    "    off = offset(1, 1, 1)\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            length = min(nx - ix, ny - iy, nz)\n",
    "            line = as_strided(grid[ix, iy, :], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for ix in range(nx):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx - ix, ny, nz - iz)\n",
    "            line = as_strided(grid[ix, :, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for iy in range(1, ny):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx, ny - iy, nz - iz)\n",
    "            line = as_strided(grid[:, iy, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # space-diagonal 2 (-1,1,1)\n",
    "    off = offset(-1, 1, 1)\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            length = min(ix + 1, ny - iy, nz)\n",
    "            line = as_strided(grid[ix, iy, :], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for ix in range(nx):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(ix + 1, ny, nz - iz)\n",
    "            line = as_strided(grid[ix, :, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for iy in range(1, ny):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx, ny - iy, nz - iz)\n",
    "            line = as_strided(grid[::-1, iy, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # space-diagonal 3 (1,-1,1)\n",
    "    off = offset(1, -1, 1)\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            length = min(nx - ix, iy + 1, nz)\n",
    "            line = as_strided(grid[ix, iy, :], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for ix in range(nx):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx - ix, ny, nz - iz)\n",
    "            line = as_strided(grid[ix, ::-1, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for iy in range(ny - 1):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx, iy + 1, nz - iz)\n",
    "            line = as_strided(grid[:, iy, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # space-diagonal 4 (-1,-1,1) equiv. to (1,1,-1)\n",
    "    off = offset(-1, -1, 1)\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            length = min(ix + 1, iy + 1, nz)\n",
    "            line = as_strided(grid[ix, iy, :], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for ix in range(nx):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(ix + 1, ny, nz - iz)\n",
    "            line = as_strided(grid[ix, ::-1, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for iy in range(ny - 1):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx, iy + 1, nz - iz)\n",
    "            line = as_strided(grid[::-1, iy, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "\n",
    "def _analyze_line(line, protein_flag):\n",
    "    \"\"\"\n",
    "    Analyze line in grid.\n",
    "    Analyze a line from the masked grid and increment\n",
    "    all ligsite scores for points between two protein grid points\n",
    "\n",
    "    assumes \"line\" to be an integer or float numpy array\n",
    "\n",
    "    protein_flag ... value to denote protein grid points\n",
    "    \"\"\"\n",
    "    # indices of the protein grid points\n",
    "    ind = np.argwhere(line == protein_flag).flatten()\n",
    "    if ind.shape[0] > 0:  # if there are any protein points in the line\n",
    "        start = int(ind[0]) + 1\n",
    "        end = int(ind[-1])\n",
    "        if (end - start) > 0:\n",
    "            # part of the line between the first and the last protein grid point\n",
    "            tmp = line[start:end]\n",
    "            # increment all non-protein grid points in that part\n",
    "            tmp[tmp > protein_flag] += 1\n",
    "\n",
    "\n",
    "def find_cavities(ligsite_grid, cutoff, gap, min_size, max_size, radius=1.4, vol_res=3.):\n",
    "    \"\"\"\n",
    "    Find cavities in a LigSite grid using sets\n",
    "    \"\"\"\n",
    "    cavities = []  # list of cavity objects to be returned\n",
    "\n",
    "    d_grid = ligsite_grid.d  # get grid spacing\n",
    "    origin = ligsite_grid.origin  # get origin of the grid in cartesian coordinates\n",
    "    grid = ligsite_grid.get_grid()  # get underlying grid array\n",
    "\n",
    "    # array of indices grid points above threshold\n",
    "    indices = np.argwhere(grid >= cutoff)\n",
    "\n",
    "    # set of those indices\n",
    "    ind_set = {tuple(row) for row in indices}\n",
    "\n",
    "    # get relative indices around (but not including) the origin\n",
    "    environment = np.array(\n",
    "        [\n",
    "            xyz\n",
    "            for xyz in itertools.product(range(-1 - gap, 2 + gap), repeat=3)\n",
    "            if xyz != (0, 0, 0)\n",
    "        ]\n",
    "    ).reshape(1, -1, 3)\n",
    "\n",
    "    while ind_set:\n",
    "        points = []  # indices of a new cavity\n",
    "\n",
    "        seeds = [\n",
    "            ind_set.pop(),\n",
    "        ]\n",
    "        while seeds:\n",
    "            # add the point(s) to the growing cavity\n",
    "            points.extend(seeds)\n",
    "\n",
    "            # create a set of neighbours\n",
    "            neighbour_coords = (\n",
    "                np.array(seeds).reshape(-1, 1, 3) + environment\n",
    "            ).reshape(-1, 3)\n",
    "            neighbours = {tuple(row) for row in neighbour_coords}\n",
    "\n",
    "            # set of neighbours that qualify as cavity points\n",
    "            in_cavity = ind_set.intersection(neighbours)\n",
    "\n",
    "            # remove them from the set of indices\n",
    "            ind_set.difference_update(in_cavity)\n",
    "\n",
    "            # these points are the 'seeds' for the next iteration\n",
    "            seeds = list(in_cavity)\n",
    "\n",
    "        # now 'points' contains a complete cavity\n",
    "        if min_size <= len(points) <= max_size:\n",
    "            cavities.append(_construct_cavity(points, grid, d_grid, origin, radius, vol_res))\n",
    "\n",
    "    # sort cavities according to their size in descending order and generate PDBStructure objects\n",
    "    cavities.sort(key=lambda x: x.size, reverse=True)\n",
    "    _gen_cav_pdb(cavities)\n",
    "\n",
    "    return cavities\n",
    "\n",
    "\n",
    "def find_cavities_dist(ligsite_grid, cutoff, max_dist, min_size, max_size, radius=1.4, vol_res=3.):\n",
    "    \"\"\"\n",
    "    Find cavities in a LigSite grid, using a distance cutoff\n",
    "    \"\"\"\n",
    "    cavities = []  # list of cavity objects to be returned\n",
    "\n",
    "    \n",
    "    d_grid = ligsite_grid.d  # get grid spacing\n",
    "    origin = ligsite_grid.origin  # get origin of the grid in cartesian coordinates\n",
    "    grid = ligsite_grid.get_grid()  # get underlying grid array\n",
    "\n",
    "    # array of indices with LigSite scores above cutoff\n",
    "    indices = np.argwhere(grid >= cutoff)\n",
    "    n_ind = indices.shape[0]\n",
    "    # flag for cavity search, True ... grid point not yet analyzed\n",
    "    flag = np.ones(n_ind, dtype=bool)\n",
    "\n",
    "    # pairwise squared distances in grid coordinates\n",
    "    norm = (indices * indices).sum(axis=1).reshape(1, -1)\n",
    "    dist = (norm + norm.T) - 2 * np.dot(indices, indices.T)\n",
    "\n",
    "    # squared distance cutoff in grid coordinates\n",
    "    sq_max = max_dist**2 / d_grid**2\n",
    "\n",
    "    for i in range(n_ind):  # step through the array by lines\n",
    "        if flag[i]:  # if we have not yet dealt with this grid point\n",
    "            seeds = [\n",
    "                i,\n",
    "            ]\n",
    "            flag[i] = False\n",
    "            points = []\n",
    "\n",
    "            while seeds:\n",
    "                tmp_list = []\n",
    "                for j in seeds:\n",
    "                    ind_j = tuple(indices[j])\n",
    "                    points.append(ind_j)\n",
    "                    neigh = np.argwhere((dist[j] <= sq_max) & flag)[:, 0]\n",
    "                    flag[neigh] = False\n",
    "                    tmp_list.extend(neigh)\n",
    "                seeds = tmp_list\n",
    "\n",
    "            # now 'points' contains a complete cavity\n",
    "            if min_size <= len(points) <= max_size:\n",
    "                cavities.append(_construct_cavity(points, grid, d_grid, origin, radius, vol_res))\n",
    "\n",
    "    # sort cavities according to their size in descending order and generate PDBStructure objects\n",
    "    cavities.sort(key=lambda x: x.size, reverse=True)\n",
    "    _gen_cav_pdb(cavities)\n",
    "\n",
    "    return cavities\n",
    "\n",
    "\n",
    "def _construct_cavity(points, grid, d_grid, origin, radius=1.4, vol_res=3.):\n",
    "    \"\"\"\n",
    "    construct a cavity from the assembled grid points\n",
    "    \"\"\"\n",
    "    points = np.array(points)  # grid indices\n",
    "    coords = points.astype(FP_DTYPE) * d_grid + origin  # 3D coordinates\n",
    "    # create new cavity object\n",
    "    cav = Cavity(points, coords, d_grid)\n",
    "    # estimate cavity volume\n",
    "    cav.volume = estimate_volume(coords, radius, d_grid / vol_res)\n",
    "    # add LigSite scores to annotations dictionary\n",
    "    cav.annotations[\"LIG\"] = grid[points[:, 0], points[:, 1], points[:, 2]]\n",
    "\n",
    "    return cav\n",
    "\n",
    "\n",
    "def _gen_cav_pdb(cavities):\n",
    "    \"\"\"\n",
    "    generate PDBStructure objects for the cavities\n",
    "    \"\"\"\n",
    "    n_at = 1\n",
    "    for n_cav, cav in enumerate(cavities):\n",
    "        cav.pdb = PDBStructure()\n",
    "        if n_cav >= 9999:  # just in case that there are more than 9999 cavities\n",
    "            n_cav -= 9999\n",
    "\n",
    "        for n_point, xyz in enumerate(cav.coords):\n",
    "            # generate a dummy atom\n",
    "            atom = PDBAtom(\n",
    "                f\"HETATM{n_at:5d}  XP  CAV X   1       3.500  53.900  22.400  1.00  7.00\"\n",
    "            )\n",
    "            # set the actual parameters\n",
    "            atom.x, atom.y, atom.z = xyz\n",
    "            atom.residue_number = n_cav + 1\n",
    "            # store the LigSite score in the B-factor column\n",
    "            atom.bfactor = cav.annotations[\"LIG\"][n_point]\n",
    "            # set the element to \"X\" (necessary for reading with Yasara)\n",
    "            atom.element = \"X\"\n",
    "            atom.keep_old_bfactor = atom.bfactor\n",
    "            atom.add_prop_dic(\"LIG\", atom.bfactor)\n",
    "            cav.pdb.atom.append(atom)\n",
    "            n_at += 1\n",
    "            if n_at > 99999:\n",
    "                n_at -= 99999\n",
    "\n",
    "\n",
    "class Cavity:\n",
    "    \"\"\"\n",
    "    definition of a cavity object\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, points, coords, d_grid):\n",
    "        self.size = len(points)\n",
    "        self.volume = 0.\n",
    "        self.d_grid = d_grid\n",
    "        self.points = points  # cavity points in grid coordinates\n",
    "        self.coords = coords  # cartesian cavity coordinates\n",
    "\n",
    "        self.annotations = dict()  # cavity point annotations\n",
    "        self.pdb = None  # cavity as a PDBStructure object\n",
    "\n",
    "        self.shaped = False  # indicate, whether the cavity was shaped\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.get_pdbstr()\n",
    "\n",
    "    def get_pdbstr(self, annotation=None):\n",
    "        header = self._get_pdb_header()\n",
    "        if annotation is None:  # use the standard B-factor column\n",
    "            return header + self.pdb.get_pdbstr()\n",
    "        else:  # replace the B-factor column with annotation value\n",
    "            return header + self.pdb.get_pdbstr(\"prop_dic\", annotation)\n",
    "\n",
    "    def _get_pdb_header(self):\n",
    "        header = [\n",
    "            \"REMARK\",\n",
    "            f\"REMARK number of grid_points:{self.size:7d}\",\n",
    "            f\"REMARK approximate cavity volume:{self.volume:7.0f} Angs.**3\",\n",
    "        ]\n",
    "        return \"\\n\".join(header) + \"\\n\"\n",
    "\n",
    "    def write(self, file_obj, file_format=\"pdb\", annotation=None):\n",
    "        if file_format == \"pdb\":\n",
    "            file_obj.write(self.get_pdbstr(annotation))\n",
    "        elif file_format == \"csv\":\n",
    "            scores = [\n",
    "                self.annotations[key] for key in annotation if key in self.annotations\n",
    "            ]\n",
    "            for i, (x, y, z) in enumerate(self.coords):\n",
    "                data = [x, y, z]\n",
    "                data.extend([score[i] for score in scores])\n",
    "                line = \", \".join([f\"{item:f}\" for item in data])\n",
    "                file_obj.write(f\"{line}\\n\")\n",
    "\n",
    "    def annotate(self, atom_coords, chg, hp, c1=1.0, c2=4.0):\n",
    "        \"\"\"\n",
    "        Annotate cavity points with Coulomb and hydrophobic potentials\n",
    "        \"\"\"\n",
    "        cp, lp = calculate_cp_lp(self.coords, atom_coords, chg, hp, c1, c2)\n",
    "        self.annotations[\"CP\"] = cp\n",
    "        self.annotations[\"HP\"] = lp\n",
    "\n",
    "        # write value to PDBatom properties 'CP' and 'HP'\n",
    "        for i, (cp_value, lp_value) in enumerate(zip(cp, lp)):\n",
    "            self.pdb.atom[i].set_prop(\"HP\", lp_value)\n",
    "            self.pdb.atom[i].add_prop_dic(\"HP\", lp_value)\n",
    "            self.pdb.atom[i].add_prop_dic(\"CP\", cp_value)\n",
    "\n",
    "    def shape(self, shaping_obj, d_max):\n",
    "        \"\"\"\n",
    "        Shape a cavity based on points in 'shaping_obj' and a maximum distance of 'd_max'.\n",
    "        The function modifies the original cavity!\n",
    "\n",
    "        :param d_max: maximum distance of a cavity point from any point in 'shaping_obj' to be retained\n",
    "        :param shaping_obj: Nx3 numpy array, coordinates of 'shaping_obj'\n",
    "        :return: number of retained cavity points\n",
    "        \"\"\"\n",
    "        flags = crop_pointcloud(self.coords, shaping_obj.astype(FP_DTYPE), d_max)\n",
    "\n",
    "        num_retained = np.sum(flags)  # number of retained points\n",
    "        self.size = num_retained\n",
    "\n",
    "        self.points = self.points[flags]\n",
    "        self.coords = self.coords[flags]\n",
    "\n",
    "        for key, value in self.annotations.items():\n",
    "            self.annotations[key] = self.annotations[key][flags]\n",
    "\n",
    "        atom = self.pdb.atom\n",
    "        self.pdb.atom = [atom[i] for i, flag in enumerate(flags) if flag]\n",
    "\n",
    "        self.shaped = True\n",
    "\n",
    "        return num_retained\n",
    "\n",
    "\n",
    "def calc_dist(cloud_1, cloud_2, do_sqrt=False):\n",
    "    \"\"\"\n",
    "    calculate the distance matrix for two point clouds\n",
    "    :param cloud_1: N x k numpy array\n",
    "    :param cloud_2: M x k numpy array\n",
    "           (k ... dimension of the clouds)\n",
    "    :param do_sqrt: if True return the square root\n",
    "    :return: N x M numpy array with euclidean distances\n",
    "    \"\"\"\n",
    "    norm_1 = (cloud_1 * cloud_1).sum(axis=1).reshape(-1, 1)\n",
    "    norm_2 = (cloud_2 * cloud_2).sum(axis=1).reshape(1, -1)\n",
    "    dist = (norm_1 + norm_2) - 2.0 * np.dot(cloud_1, cloud_2.T)\n",
    "\n",
    "    if do_sqrt:\n",
    "        return np.sqrt(dist, out=dist)\n",
    "    else:\n",
    "        return dist\n",
    "\n",
    "\n",
    "def crop_pointcloud(cloud_1, cloud_2, d_max=2.0):\n",
    "    \"\"\"\n",
    "    crop one point cloud based on the points of another point cloud,\n",
    "    all points in 'big', which are closer than 'd_max' to any point in 'small',\n",
    "    are retained\n",
    "\n",
    "    :param cloud_1: N x k numpy array\n",
    "    :param cloud_2: M x k numpy array\n",
    "                  (k ... dimension of the clouds)\n",
    "    :param d_max: maximum distance for a point in 'cloud_1' to be retained\n",
    "    :return: boolean array with shape (N,)\n",
    "             True ... point is retained\n",
    "             False ... point is deleted\n",
    "    \"\"\"\n",
    "    diff = calc_dist(cloud_1, cloud_2)  # calculate all pairwise distances\n",
    "    min_dist = diff.min(axis=1)  # get the minimum distance\n",
    "\n",
    "    return min_dist <= d_max**2\n",
    "\n",
    "\n",
    "def calculate_cp_lp(points, atom_coords, chg, hp, c1=1.0, c2=4.0):\n",
    "    \"\"\"\n",
    "    Calculate the Coulomb and hydrophobic potential at points using the coordinates\n",
    "    in 'atom_coords' and the (partial) charges and atomic hydrophobicity values in 'chg' and 'hp'.\n",
    "    Uses Fermi-weighting for calculating the hydrophobic potential and a relative epsilon of 'r'.\n",
    "\n",
    "    :param points: points at the which the potentials will be calculated (M, 3)\n",
    "    :param atom_coords: protein coordinates (N, 3)\n",
    "    :param chg: (partial) charges assigned to the protein atoms (N,)\n",
    "    :param hp: hydrophobicity parameters assigned to the protein atoms (N,)\n",
    "    :param c1: parameter for Fermi-weighting\n",
    "    :param c2: parameter for Fermi-weighting\n",
    "    :return: cp, lp (M,)\n",
    "    \"\"\"\n",
    "    # calculate squared distances\n",
    "    dist_sq = calc_dist(points, atom_coords, do_sqrt=False)\n",
    "\n",
    "    # Coulomb potential, relative permittivity = distance\n",
    "    # cp = 557.0 * sum(chg / dist**2)\n",
    "    cp = 557.0 * np.sum(np.divide(chg[np.newaxis, :], dist_sq), axis=1)\n",
    "\n",
    "    # calculate distances, dist_sq is overwritten!\n",
    "    dist = np.sqrt(dist_sq, out=dist_sq)\n",
    "\n",
    "    # fermi_wt = (math.exp(-c1 * c2) + 1.0) / (np.exp(c1 * (dist - c2)) + 1.0)\n",
    "    # calculation of the weighting term is spilt-up and uses\n",
    "    # the 'out' parameter to prevent creation of a new array at each step\n",
    "    fermi_wt = dist  # COPY if 'dist' is used later\n",
    "    np.subtract(fermi_wt, c2, out=fermi_wt)  # dist - c2\n",
    "    np.multiply(fermi_wt, c1, out=fermi_wt)  # c1 * (dist - c2)\n",
    "    if FP_DTYPE == np.float32:\n",
    "        # clipping is necessary to prevent overflow in np.exp\n",
    "        np.clip(fermi_wt, -103.9, 88.7, out=fermi_wt)\n",
    "    np.exp(fermi_wt, out=fermi_wt)  # np.exp(c1 * (dist - c2))\n",
    "    np.add(fermi_wt, 1.0, out=fermi_wt)  # np.exp(c1 * (dist - c2)) + 1.0\n",
    "    # (math.exp(-c1 * c2) + 1.0) / (np.exp(c1 * (dist - c2)) + 1.0)\n",
    "    np.divide(math.exp(-c1 * c2) + 1.0, fermi_wt, out=fermi_wt)\n",
    "\n",
    "    # sum of weights, clipped to 1.e-6\n",
    "    # MUST be calculated first, because the next step overwrites 'fermi_wt'\n",
    "    sum_wt = np.clip(np.sum(fermi_wt, axis=1), 1.0e-6, np.inf)\n",
    "    # weighted sum of atom contributions\n",
    "    lp = np.sum(np.multiply(fermi_wt, hp[np.newaxis, :], out=fermi_wt), axis=1)\n",
    "\n",
    "    np.divide(lp, sum_wt, out=lp)\n",
    "\n",
    "    return cp, lp\n",
    "\n",
    "\n",
    "def estimate_volume(points, radius: float, d_grid: float) -> float:\n",
    "    \"\"\"\n",
    "    estimate the volume of a point cloud\n",
    "    point ... 3D coordinates of the cloud\n",
    "    radius ... radius of the sphere placed at each point\n",
    "    d_grid ... grid spacing\n",
    "\n",
    "    The object is placed on a regular grid; voxels within the spheres placed at the points are\n",
    "    counted; the estimated volume is n_voxels * d_grid**3\n",
    "    \"\"\"\n",
    "    radii = np.ones(points.shape[0], dtype=FP_DTYPE) * radius\n",
    "    grid = setup_grid(points, radius + d_grid, d_grid, 0, np.int8)\n",
    "    mask_grid(grid, points, radii, 1.0, 0.0, 0.0, 1, 0)\n",
    "    return np.sum(grid.get_grid()) * d_grid ** 3\n",
    "\n",
    "\n",
    "class Grid:\n",
    "    \"\"\"\n",
    "    Class for a grid object.\n",
    "\n",
    "    Extension (# of grid points) given by a tuple of\n",
    "    length 3. Default initialization with value \"0\".\n",
    "\n",
    "    USAGE: Grid(origin, extend, init=0)\n",
    "              extend: tuple:(x,y,z)  grid extension number of grid points\n",
    "                              in (x , y , z)\n",
    "              init:        :initialization value of grid point (standard=0)\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self, origin=(0.0, 0.0, 0.0), extent=(50, 50, 50), d=0.7, init=0, dtype=np.int32\n",
    "    ):\n",
    "        self.nx, self.ny, self.nz = extent\n",
    "        self.extent = extent\n",
    "        self.origin = np.array(origin, dtype=FP_DTYPE)\n",
    "        self.d = d\n",
    "        self._grid = np.zeros(extent, dtype=dtype)\n",
    "        if init != 0:\n",
    "            self._grid = init\n",
    "\n",
    "    def get_subgrid(self, x0, y0, z0, x1, y1, z1):\n",
    "        \"\"\"\n",
    "        Return a view of a sub-grid of the internal np.array\n",
    "        :param x0: start x-index\n",
    "        :param y0: start y-index\n",
    "        :param z0: start z-index\n",
    "        :param x1: end x-index\n",
    "        :param y1: end y-index\n",
    "        :param z1: end z-index\n",
    "        :return: reference to the sub-grid\n",
    "        \"\"\"\n",
    "        return self._grid[x0:x1, y0:y1, z0:z1]\n",
    "\n",
    "    def get_grid(self):\n",
    "        \"\"\"\n",
    "        Return a view of the complete grid object\n",
    "        :return: reference to the internal np.array\n",
    "        \"\"\"\n",
    "        return self._grid\n",
    "\n",
    "    def get_value(self, indices):\n",
    "        \"\"\"\n",
    "        Return value of grid point\n",
    "        :param indices: tuple of indices (i, j, k); cannot be a np.array\n",
    "        :return: value at grid point\n",
    "        \"\"\"\n",
    "        return self._grid[indices]\n",
    "\n",
    "    def set_value(self, indices, value):\n",
    "        \"\"\"\n",
    "        Set value of a grid point\n",
    "        :param indices: tuple of indices (i, j, k); cannot be a np.array\n",
    "        :param value: new value\n",
    "        \"\"\"\n",
    "        self._grid[indices] = value\n",
    "\n",
    "    def is_valid_index(self, index):\n",
    "        \"\"\"\n",
    "        Check whether 'index' contains valid indices of the grid.\n",
    "        :param index: tuple of indices (i, j, k); cannot be an np.array\n",
    "        :return: True/False\n",
    "        \"\"\"\n",
    "        x, y, z = index\n",
    "        return 0 <= x < self.nx and 0 <= y < self.ny and 0 <= z < self.nz\n",
    "\n",
    "    def coordinates(self, index):\n",
    "        \"\"\"\n",
    "        Return the cartesian coordinates of a grid point\n",
    "        :param index: indices (i, j, k), tuple, list, np.array\n",
    "        :return: np.array with cartesian coordinates\n",
    "        \"\"\"\n",
    "        return self.origin + np.array(index, dtype=FP_DTYPE) * self.d\n",
    "\n",
    "    def offset(self, dx, dy, dz):\n",
    "        \"\"\"\n",
    "        Return the offset in bytes\n",
    "        :param dx: increment in x\n",
    "        :param dy: increment in y\n",
    "        :param dz: increment in z\n",
    "        :return: offset, to be used for as_strided\n",
    "        \"\"\"\n",
    "        return ((dx * self.ny + dy) * self.nz + dz) * self._grid.itemsize\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2cf1ccaf-1e01-4b8a-8772-7ef221699fad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "function definitions for LigSite and cavity detection\n",
    "\"\"\"\n",
    "import itertools\n",
    "import math\n",
    "\n",
    "import numpy as np\n",
    "from numpy.lib.stride_tricks import as_strided\n",
    "\n",
    "\n",
    "# 32-bit precision provides speed-up in HP annotation\n",
    "FP_DTYPE = np.float32\n",
    "\n",
    "\n",
    "def setup_grid(coords, cushion, d_grid, init, dtype):\n",
    "    \"\"\"\n",
    "    Setup and initialize grid using atom coordinates\n",
    "    \"\"\"\n",
    "    # min and max in cartesian coordinates\n",
    "    min_coords = coords.min(axis=0)\n",
    "    max_coords = coords.max(axis=0)\n",
    "    # min and max in grid coordinates\n",
    "    min_grid = np.floor((min_coords - cushion) / d_grid)\n",
    "    max_grid = np.ceil((max_coords + cushion) / d_grid)\n",
    "\n",
    "    origin = min_grid * d_grid\n",
    "    extent = (max_grid - min_grid + 1).astype(int)\n",
    "\n",
    "    return Grid(origin=origin, extent=extent, d=d_grid, init=init, dtype=dtype)\n",
    "\n",
    "\n",
    "def mask_grid(\n",
    "    grid, coords, radii, radius_factor, probe_radius, softness, protein_flag, soft_flag\n",
    "):\n",
    "    \"\"\"\n",
    "    Mask grid using atom coordinates and radii\n",
    "    \"\"\"\n",
    "    # atom coordinates in grid units\n",
    "    grid_coords = (coords - grid.origin) / grid.d\n",
    "    # atom radii in grid units\n",
    "    grid_radii = (radii * radius_factor + probe_radius) / grid.d\n",
    "    r1sq = grid_radii**2  # squared outer, soft radius\n",
    "    r2sq = (grid_radii - softness) ** 2  # squared inner, hard radius\n",
    "\n",
    "    # origin and space-diagonal coordinates of the sub-grids around the atoms\n",
    "    sg_start = np.clip(\n",
    "        np.floor(grid_coords - grid_radii.reshape(-1, 1)).astype(int),\n",
    "        (0, 0, 0),\n",
    "        grid.extent,\n",
    "    )\n",
    "    sg_end = np.clip(\n",
    "        np.ceil(grid_coords + grid_radii.reshape(-1, 1)).astype(int) + 1,\n",
    "        (0, 0, 0),\n",
    "        grid.extent,\n",
    "    )\n",
    "\n",
    "    for i in range(grid_coords.shape[0]):  # loop over all atom coordinates\n",
    "        x_start, y_start, z_start = sg_start[i]\n",
    "        x_end, y_end, z_end = sg_end[i]\n",
    "        sub_grid = grid.get_subgrid(x_start, y_start, z_start, x_end, y_end, z_end)\n",
    "\n",
    "        x, y, z = np.ogrid[x_start:x_end, y_start:y_end, z_start:z_end]\n",
    "        dist = (\n",
    "            (x - grid_coords[i, 0]) ** 2\n",
    "            + (y - grid_coords[i, 1]) ** 2\n",
    "            + (z - grid_coords[i, 2]) ** 2\n",
    "        )\n",
    "\n",
    "        # grid points within the inner, hard radius\n",
    "        mask_hard = dist < r2sq[i]\n",
    "        sub_grid[mask_hard] = protein_flag\n",
    "\n",
    "        # grid points between inner and outer radius\n",
    "        mask_soft = (dist < r1sq[i]) & (sub_grid != protein_flag)\n",
    "        sub_grid[mask_soft] = soft_flag\n",
    "\n",
    "\n",
    "def do_ligsite(masked_grid, protein_flag):\n",
    "    \"\"\"\n",
    "    Run ligsite algorithm.\n",
    "    Calculate LigSite scores for a masked grid\n",
    "    \"\"\"\n",
    "    grid = masked_grid.get_grid()\n",
    "    nx = masked_grid.nx\n",
    "    ny = masked_grid.ny\n",
    "    nz = masked_grid.nz\n",
    "    offset = masked_grid.offset\n",
    "\n",
    "    # analyze along the x-direction\n",
    "    for iy in range(ny):\n",
    "        for iz in range(nz):\n",
    "            line = grid[:, iy, iz]\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # analyze along the y-direction\n",
    "    for ix in range(nx):\n",
    "        for iz in range(nz):\n",
    "            line = grid[ix, :, iz]\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # analyze along the z-direction\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            line = grid[ix, iy, :]\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # space-diagonal 1 (1,1,1)\n",
    "    off = offset(1, 1, 1)\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            length = min(nx - ix, ny - iy, nz)\n",
    "            line = as_strided(grid[ix, iy, :], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for ix in range(nx):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx - ix, ny, nz - iz)\n",
    "            line = as_strided(grid[ix, :, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for iy in range(1, ny):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx, ny - iy, nz - iz)\n",
    "            line = as_strided(grid[:, iy, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # space-diagonal 2 (-1,1,1)\n",
    "    off = offset(-1, 1, 1)\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            length = min(ix + 1, ny - iy, nz)\n",
    "            line = as_strided(grid[ix, iy, :], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for ix in range(nx):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(ix + 1, ny, nz - iz)\n",
    "            line = as_strided(grid[ix, :, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for iy in range(1, ny):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx, ny - iy, nz - iz)\n",
    "            line = as_strided(grid[::-1, iy, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # space-diagonal 3 (1,-1,1)\n",
    "    off = offset(1, -1, 1)\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            length = min(nx - ix, iy + 1, nz)\n",
    "            line = as_strided(grid[ix, iy, :], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for ix in range(nx):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx - ix, ny, nz - iz)\n",
    "            line = as_strided(grid[ix, ::-1, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for iy in range(ny - 1):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx, iy + 1, nz - iz)\n",
    "            line = as_strided(grid[:, iy, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    # space-diagonal 4 (-1,-1,1) equiv. to (1,1,-1)\n",
    "    off = offset(-1, -1, 1)\n",
    "    for ix in range(nx):\n",
    "        for iy in range(ny):\n",
    "            length = min(ix + 1, iy + 1, nz)\n",
    "            line = as_strided(grid[ix, iy, :], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for ix in range(nx):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(ix + 1, ny, nz - iz)\n",
    "            line = as_strided(grid[ix, ::-1, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "    for iy in range(ny - 1):\n",
    "        for iz in range(1, nz):\n",
    "            length = min(nx, iy + 1, nz - iz)\n",
    "            line = as_strided(grid[::-1, iy, iz], shape=(length,), strides=(off,))\n",
    "            _analyze_line(line, protein_flag)\n",
    "\n",
    "\n",
    "def _analyze_line(line, protein_flag):\n",
    "    \"\"\"\n",
    "    Analyze line in grid.\n",
    "    Analyze a line from the masked grid and increment\n",
    "    all ligsite scores for points between two protein grid points\n",
    "\n",
    "    assumes \"line\" to be an integer or float numpy array\n",
    "\n",
    "    protein_flag ... value to denote protein grid points\n",
    "    \"\"\"\n",
    "    # indices of the protein grid points\n",
    "    ind = np.argwhere(line == protein_flag).flatten()\n",
    "    if ind.shape[0] > 0:  # if there are any protein points in the line\n",
    "        start = int(ind[0]) + 1\n",
    "        end = int(ind[-1])\n",
    "        if (end - start) > 0:\n",
    "            # part of the line between the first and the last protein grid point\n",
    "            tmp = line[start:end]\n",
    "            # increment all non-protein grid points in that part\n",
    "            tmp[tmp > protein_flag] += 1\n",
    "\n",
    "\n",
    "def find_cavities(ligsite_grid, cutoff, gap, min_size, max_size, radius=1.4, vol_res=3.):\n",
    "    \"\"\"\n",
    "    Find cavities in a LigSite grid using sets\n",
    "    \"\"\"\n",
    "    cavities = []  # list of cavity objects to be returned\n",
    "\n",
    "    super(Grid, ligsite_grid).__init__\n",
    "    d_grid = ligsite_grid.d  # get grid spacing\n",
    "    origin = ligsite_grid.origin  # get origin of the grid in cartesian coordinates\n",
    "    grid = ligsite_grid.get_grid()  # get underlying grid array\n",
    "\n",
    "    # array of indices grid points above threshold\n",
    "    indices = np.argwhere(grid >= cutoff)\n",
    "\n",
    "    # set of those indices\n",
    "    ind_set = {tuple(row) for row in indices}\n",
    "\n",
    "    # get relative indices around (but not including) the origin\n",
    "    environment = np.array(\n",
    "        [\n",
    "            xyz\n",
    "            for xyz in itertools.product(range(-1 - gap, 2 + gap), repeat=3)\n",
    "            if xyz != (0, 0, 0)\n",
    "        ]\n",
    "    ).reshape(1, -1, 3)\n",
    "\n",
    "    while ind_set:\n",
    "        points = []  # indices of a new cavity\n",
    "\n",
    "        seeds = [\n",
    "            ind_set.pop(),\n",
    "        ]\n",
    "        while seeds:\n",
    "            # add the point(s) to the growing cavity\n",
    "            points.extend(seeds)\n",
    "\n",
    "            # create a set of neighbours\n",
    "            neighbour_coords = (\n",
    "                np.array(seeds).reshape(-1, 1, 3) + environment\n",
    "            ).reshape(-1, 3)\n",
    "            neighbours = {tuple(row) for row in neighbour_coords}\n",
    "\n",
    "            # set of neighbours that qualify as cavity points\n",
    "            in_cavity = ind_set.intersection(neighbours)\n",
    "\n",
    "            # remove them from the set of indices\n",
    "            ind_set.difference_update(in_cavity)\n",
    "\n",
    "            # these points are the 'seeds' for the next iteration\n",
    "            seeds = list(in_cavity)\n",
    "\n",
    "        # now 'points' contains a complete cavity\n",
    "        if min_size <= len(points) <= max_size:\n",
    "            cavities.append(_construct_cavity(points, grid, d_grid, origin, radius, vol_res))\n",
    "\n",
    "    # sort cavities according to their size in descending order and generate PDBStructure objects\n",
    "    cavities.sort(key=lambda x: x.size, reverse=True)\n",
    "    _gen_cav_pdb(cavities)\n",
    "\n",
    "    return cavities\n",
    "\n",
    "\n",
    "def find_cavities_dist(ligsite_grid, cutoff, max_dist, min_size, max_size, radius=1.4, vol_res=3.):\n",
    "    \"\"\"\n",
    "    Find cavities in a LigSite grid, using a distance cutoff\n",
    "    \"\"\"\n",
    "    cavities = []  # list of cavity objects to be returned\n",
    "\n",
    "    d_grid = ligsite_grid.d  # get grid spacing\n",
    "    origin = ligsite_grid.origin  # get origin of the grid in cartesian coordinates\n",
    "    grid = ligsite_grid.get_grid()  # get underlying grid array\n",
    "\n",
    "    # array of indices with LigSite scores above cutoff\n",
    "    indices = np.argwhere(grid >= cutoff)\n",
    "    n_ind = indices.shape[0]\n",
    "    # flag for cavity search, True ... grid point not yet analyzed\n",
    "    flag = np.ones(n_ind, dtype=bool)\n",
    "\n",
    "    # pairwise squared distances in grid coordinates\n",
    "    norm = (indices * indices).sum(axis=1).reshape(1, -1)\n",
    "    dist = (norm + norm.T) - 2 * np.dot(indices, indices.T)\n",
    "\n",
    "    # squared distance cutoff in grid coordinates\n",
    "    sq_max = max_dist**2 / d_grid**2\n",
    "\n",
    "    for i in range(n_ind):  # step through the array by lines\n",
    "        if flag[i]:  # if we have not yet dealt with this grid point\n",
    "            seeds = [\n",
    "                i,\n",
    "            ]\n",
    "            flag[i] = False\n",
    "            points = []\n",
    "\n",
    "            while seeds:\n",
    "                tmp_list = []\n",
    "                for j in seeds:\n",
    "                    ind_j = tuple(indices[j])\n",
    "                    points.append(ind_j)\n",
    "                    neigh = np.argwhere((dist[j] <= sq_max) & flag)[:, 0]\n",
    "                    flag[neigh] = False\n",
    "                    tmp_list.extend(neigh)\n",
    "                seeds = tmp_list\n",
    "\n",
    "            # now 'points' contains a complete cavity\n",
    "            if min_size <= len(points) <= max_size:\n",
    "                cavities.append(_construct_cavity(points, grid, d_grid, origin, radius, vol_res))\n",
    "\n",
    "    # sort cavities according to their size in descending order and generate PDBStructure objects\n",
    "    cavities.sort(key=lambda x: x.size, reverse=True)\n",
    "    _gen_cav_pdb(cavities)\n",
    "\n",
    "    return cavities\n",
    "\n",
    "\n",
    "def _construct_cavity(points, grid, d_grid, origin, radius=1.4, vol_res=3.):\n",
    "    \"\"\"\n",
    "    construct a cavity from the assembled grid points\n",
    "    \"\"\"\n",
    "    points = np.array(points)  # grid indices\n",
    "    coords = points.astype(FP_DTYPE) * d_grid + origin  # 3D coordinates\n",
    "    # create new cavity object\n",
    "    cav = Cavity(points, coords, d_grid)\n",
    "    # estimate cavity volume\n",
    "    cav.volume = estimate_volume(coords, radius, d_grid / vol_res)\n",
    "    # add LigSite scores to annotations dictionary\n",
    "    cav.annotations[\"LIG\"] = grid[points[:, 0], points[:, 1], points[:, 2]]\n",
    "\n",
    "    return cav\n",
    "\n",
    "\n",
    "def _gen_cav_pdb(cavities):\n",
    "    \"\"\"\n",
    "    generate PDBStructure objects for the cavities\n",
    "    \"\"\"\n",
    "    n_at = 1\n",
    "    for n_cav, cav in enumerate(cavities):\n",
    "        cav.pdb = PDBStructure()\n",
    "        if n_cav >= 9999:  # just in case that there are more than 9999 cavities\n",
    "            n_cav -= 9999\n",
    "\n",
    "        for n_point, xyz in enumerate(cav.coords):\n",
    "            # generate a dummy atom\n",
    "            atom = PDBAtom(\n",
    "                f\"HETATM{n_at:5d}  XP  CAV X   1       3.500  53.900  22.400  1.00  7.00\"\n",
    "            )\n",
    "            # set the actual parameters\n",
    "            atom.x, atom.y, atom.z = xyz\n",
    "            atom.residue_number = n_cav + 1\n",
    "            # store the LigSite score in the B-factor column\n",
    "            atom.bfactor = cav.annotations[\"LIG\"][n_point]\n",
    "            # set the element to \"X\" (necessary for reading with Yasara)\n",
    "            atom.element = \"X\"\n",
    "            atom.keep_old_bfactor = atom.bfactor\n",
    "            atom.add_prop_dic(\"LIG\", atom.bfactor)\n",
    "            cav.pdb.atom.append(atom)\n",
    "            n_at += 1\n",
    "            if n_at > 99999:\n",
    "                n_at -= 99999\n",
    "\n",
    "\n",
    "class Cavity:\n",
    "    \"\"\"\n",
    "    definition of a cavity object\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, points, coords, d_grid):\n",
    "        self.size = len(points)\n",
    "        self.volume = 0.\n",
    "        self.d_grid = d_grid\n",
    "        self.points = points  # cavity points in grid coordinates\n",
    "        self.coords = coords  # cartesian cavity coordinates\n",
    "\n",
    "        self.annotations = dict()  # cavity point annotations\n",
    "        self.pdb = None  # cavity as a PDBStructure object\n",
    "\n",
    "        self.shaped = False  # indicate, whether the cavity was shaped\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.get_pdbstr()\n",
    "\n",
    "    def get_pdbstr(self, annotation=None):\n",
    "        header = self._get_pdb_header()\n",
    "        if annotation is None:  # use the standard B-factor column\n",
    "            return header + self.pdb.get_pdbstr()\n",
    "        else:  # replace the B-factor column with annotation value\n",
    "            return header + self.pdb.get_pdbstr(\"prop_dic\", annotation)\n",
    "\n",
    "    def _get_pdb_header(self):\n",
    "        header = [\n",
    "            \"REMARK\",\n",
    "            f\"REMARK number of grid_points:{self.size:7d}\",\n",
    "            f\"REMARK approximate cavity volume:{self.volume:7.0f} Angs.**3\",\n",
    "        ]\n",
    "        return \"\\n\".join(header) + \"\\n\"\n",
    "\n",
    "    def write(self, file_obj, file_format=\"pdb\", annotation=None):\n",
    "        if file_format == \"pdb\":\n",
    "            file_obj.write(self.get_pdbstr(annotation))\n",
    "        elif file_format == \"csv\":\n",
    "            scores = [\n",
    "                self.annotations[key] for key in annotation if key in self.annotations\n",
    "            ]\n",
    "            for i, (x, y, z) in enumerate(self.coords):\n",
    "                data = [x, y, z]\n",
    "                data.extend([score[i] for score in scores])\n",
    "                line = \", \".join([f\"{item:f}\" for item in data])\n",
    "                file_obj.write(f\"{line}\\n\")\n",
    "\n",
    "    def annotate(self, atom_coords, chg, hp, c1=1.0, c2=4.0):\n",
    "        \"\"\"\n",
    "        Annotate cavity points with Coulomb and hydrophobic potentials\n",
    "        \"\"\"\n",
    "        cp, lp = calculate_cp_lp(self.coords, atom_coords, chg, hp, c1, c2)\n",
    "        self.annotations[\"CP\"] = cp\n",
    "        self.annotations[\"HP\"] = lp\n",
    "\n",
    "        # write value to PDBatom properties 'CP' and 'HP'\n",
    "        for i, (cp_value, lp_value) in enumerate(zip(cp, lp)):\n",
    "            self.pdb.atom[i].set_prop(\"HP\", lp_value)\n",
    "            self.pdb.atom[i].add_prop_dic(\"HP\", lp_value)\n",
    "            self.pdb.atom[i].add_prop_dic(\"CP\", cp_value)\n",
    "\n",
    "    def shape(self, shaping_obj, d_max):\n",
    "        \"\"\"\n",
    "        Shape a cavity based on points in 'shaping_obj' and a maximum distance of 'd_max'.\n",
    "        The function modifies the original cavity!\n",
    "\n",
    "        :param d_max: maximum distance of a cavity point from any point in 'shaping_obj' to be retained\n",
    "        :param shaping_obj: Nx3 numpy array, coordinates of 'shaping_obj'\n",
    "        :return: number of retained cavity points\n",
    "        \"\"\"\n",
    "        flags = crop_pointcloud(self.coords, shaping_obj.astype(FP_DTYPE), d_max)\n",
    "\n",
    "        num_retained = np.sum(flags)  # number of retained points\n",
    "        self.size = num_retained\n",
    "\n",
    "        self.points = self.points[flags]\n",
    "        self.coords = self.coords[flags]\n",
    "\n",
    "        for key, value in self.annotations.items():\n",
    "            self.annotations[key] = self.annotations[key][flags]\n",
    "\n",
    "        atom = self.pdb.atom\n",
    "        self.pdb.atom = [atom[i] for i, flag in enumerate(flags) if flag]\n",
    "\n",
    "        self.shaped = True\n",
    "\n",
    "        return num_retained\n",
    "\n",
    "\n",
    "def calc_dist(cloud_1, cloud_2, do_sqrt=False):\n",
    "    \"\"\"\n",
    "    calculate the distance matrix for two point clouds\n",
    "    :param cloud_1: N x k numpy array\n",
    "    :param cloud_2: M x k numpy array\n",
    "           (k ... dimension of the clouds)\n",
    "    :param do_sqrt: if True return the square root\n",
    "    :return: N x M numpy array with euclidean distances\n",
    "    \"\"\"\n",
    "    norm_1 = (cloud_1 * cloud_1).sum(axis=1).reshape(-1, 1)\n",
    "    norm_2 = (cloud_2 * cloud_2).sum(axis=1).reshape(1, -1)\n",
    "    dist = (norm_1 + norm_2) - 2.0 * np.dot(cloud_1, cloud_2.T)\n",
    "\n",
    "    if do_sqrt:\n",
    "        return np.sqrt(dist, out=dist)\n",
    "    else:\n",
    "        return dist\n",
    "\n",
    "\n",
    "def crop_pointcloud(cloud_1, cloud_2, d_max=2.0):\n",
    "    \"\"\"\n",
    "    crop one point cloud based on the points of another point cloud,\n",
    "    all points in 'big', which are closer than 'd_max' to any point in 'small',\n",
    "    are retained\n",
    "\n",
    "    :param cloud_1: N x k numpy array\n",
    "    :param cloud_2: M x k numpy array\n",
    "                  (k ... dimension of the clouds)\n",
    "    :param d_max: maximum distance for a point in 'cloud_1' to be retained\n",
    "    :return: boolean array with shape (N,)\n",
    "             True ... point is retained\n",
    "             False ... point is deleted\n",
    "    \"\"\"\n",
    "    diff = calc_dist(cloud_1, cloud_2)  # calculate all pairwise distances\n",
    "    min_dist = diff.min(axis=1)  # get the minimum distance\n",
    "\n",
    "    return min_dist <= d_max**2\n",
    "\n",
    "\n",
    "def calculate_cp_lp(points, atom_coords, chg, hp, c1=1.0, c2=4.0):\n",
    "    \"\"\"\n",
    "    Calculate the Coulomb and hydrophobic potential at points using the coordinates\n",
    "    in 'atom_coords' and the (partial) charges and atomic hydrophobicity values in 'chg' and 'hp'.\n",
    "    Uses Fermi-weighting for calculating the hydrophobic potential and a relative epsilon of 'r'.\n",
    "\n",
    "    :param points: points at the which the potentials will be calculated (M, 3)\n",
    "    :param atom_coords: protein coordinates (N, 3)\n",
    "    :param chg: (partial) charges assigned to the protein atoms (N,)\n",
    "    :param hp: hydrophobicity parameters assigned to the protein atoms (N,)\n",
    "    :param c1: parameter for Fermi-weighting\n",
    "    :param c2: parameter for Fermi-weighting\n",
    "    :return: cp, lp (M,)\n",
    "    \"\"\"\n",
    "    # calculate squared distances\n",
    "    dist_sq = calc_dist(points, atom_coords, do_sqrt=False)\n",
    "\n",
    "    # Coulomb potential, relative permittivity = distance\n",
    "    # cp = 557.0 * sum(chg / dist**2)\n",
    "    cp = 557.0 * np.sum(np.divide(chg[np.newaxis, :], dist_sq), axis=1)\n",
    "\n",
    "    # calculate distances, dist_sq is overwritten!\n",
    "    dist = np.sqrt(dist_sq, out=dist_sq)\n",
    "\n",
    "    # fermi_wt = (math.exp(-c1 * c2) + 1.0) / (np.exp(c1 * (dist - c2)) + 1.0)\n",
    "    # calculation of the weighting term is spilt-up and uses\n",
    "    # the 'out' parameter to prevent creation of a new array at each step\n",
    "    fermi_wt = dist  # COPY if 'dist' is used later\n",
    "    np.subtract(fermi_wt, c2, out=fermi_wt)  # dist - c2\n",
    "    np.multiply(fermi_wt, c1, out=fermi_wt)  # c1 * (dist - c2)\n",
    "    if FP_DTYPE == np.float32:\n",
    "        # clipping is necessary to prevent overflow in np.exp\n",
    "        np.clip(fermi_wt, -103.9, 88.7, out=fermi_wt)\n",
    "    np.exp(fermi_wt, out=fermi_wt)  # np.exp(c1 * (dist - c2))\n",
    "    np.add(fermi_wt, 1.0, out=fermi_wt)  # np.exp(c1 * (dist - c2)) + 1.0\n",
    "    # (math.exp(-c1 * c2) + 1.0) / (np.exp(c1 * (dist - c2)) + 1.0)\n",
    "    np.divide(math.exp(-c1 * c2) + 1.0, fermi_wt, out=fermi_wt)\n",
    "\n",
    "    # sum of weights, clipped to 1.e-6\n",
    "    # MUST be calculated first, because the next step overwrites 'fermi_wt'\n",
    "    sum_wt = np.clip(np.sum(fermi_wt, axis=1), 1.0e-6, np.inf)\n",
    "    # weighted sum of atom contributions\n",
    "    lp = np.sum(np.multiply(fermi_wt, hp[np.newaxis, :], out=fermi_wt), axis=1)\n",
    "\n",
    "    np.divide(lp, sum_wt, out=lp)\n",
    "\n",
    "    return cp, lp\n",
    "\n",
    "\n",
    "def estimate_volume(points, radius: float, d_grid: float) -> float:\n",
    "    \"\"\"\n",
    "    estimate the volume of a point cloud\n",
    "    point ... 3D coordinates of the cloud\n",
    "    radius ... radius of the sphere placed at each point\n",
    "    d_grid ... grid spacing\n",
    "\n",
    "    The object is placed on a regular grid; voxels within the spheres placed at the points are\n",
    "    counted; the estimated volume is n_voxels * d_grid**3\n",
    "    \"\"\"\n",
    "    radii = np.ones(points.shape[0], dtype=FP_DTYPE) * radius\n",
    "    grid = setup_grid(points, radius + d_grid, d_grid, 0, np.int8)\n",
    "    mask_grid(grid, points, radii, 1.0, 0.0, 0.0, 1, 0)\n",
    "    return np.sum(grid.get_grid()) * d_grid ** 3\n",
    "\n",
    "\n",
    "class Grid:\n",
    "    \"\"\"\n",
    "    Class for a grid object.\n",
    "\n",
    "    Extension (# of grid points) given by a tuple of\n",
    "    length 3. Default initialization with value \"0\".\n",
    "\n",
    "    USAGE: Grid(origin, extend, init=0)\n",
    "              extend: tuple:(x,y,z)  grid extension number of grid points\n",
    "                              in (x , y , z)\n",
    "              init:        :initialization value of grid point (standard=0)\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self, origin=(0.0, 0.0, 0.0), extent=(50, 50, 50), d=0.7, init=0, dtype=np.int32\n",
    "    ):\n",
    "        self.nx, self.ny, self.nz = extent\n",
    "        self.extent = extent\n",
    "        self.origin = np.array(origin, dtype=FP_DTYPE)\n",
    "        self.d = d\n",
    "        self._grid = np.zeros(extent, dtype=dtype)\n",
    "        if init != 0:\n",
    "            self._grid = init\n",
    "\n",
    "    def get_subgrid(self, x0, y0, z0, x1, y1, z1):\n",
    "        \"\"\"\n",
    "        Return a view of a sub-grid of the internal np.array\n",
    "        :param x0: start x-index\n",
    "        :param y0: start y-index\n",
    "        :param z0: start z-index\n",
    "        :param x1: end x-index\n",
    "        :param y1: end y-index\n",
    "        :param z1: end z-index\n",
    "        :return: reference to the sub-grid\n",
    "        \"\"\"\n",
    "        return self._grid[x0:x1, y0:y1, z0:z1]\n",
    "\n",
    "    def get_grid(self):\n",
    "        \"\"\"\n",
    "        Return a view of the complete grid object\n",
    "        :return: reference to the internal np.array\n",
    "        \"\"\"\n",
    "        return self._grid\n",
    "\n",
    "    def get_value(self, indices):\n",
    "        \"\"\"\n",
    "        Return value of grid point\n",
    "        :param indices: tuple of indices (i, j, k); cannot be a np.array\n",
    "        :return: value at grid point\n",
    "        \"\"\"\n",
    "        return self._grid[indices]\n",
    "\n",
    "    def set_value(self, indices, value):\n",
    "        \"\"\"\n",
    "        Set value of a grid point\n",
    "        :param indices: tuple of indices (i, j, k); cannot be a np.array\n",
    "        :param value: new value\n",
    "        \"\"\"\n",
    "        self._grid[indices] = value\n",
    "\n",
    "    def is_valid_index(self, index):\n",
    "        \"\"\"\n",
    "        Check whether 'index' contains valid indices of the grid.\n",
    "        :param index: tuple of indices (i, j, k); cannot be an np.array\n",
    "        :return: True/False\n",
    "        \"\"\"\n",
    "        x, y, z = index\n",
    "        return 0 <= x < self.nx and 0 <= y < self.ny and 0 <= z < self.nz\n",
    "\n",
    "    def coordinates(self, index):\n",
    "        \"\"\"\n",
    "        Return the cartesian coordinates of a grid point\n",
    "        :param index: indices (i, j, k), tuple, list, np.array\n",
    "        :return: np.array with cartesian coordinates\n",
    "        \"\"\"\n",
    "        return self.origin + np.array(index, dtype=FP_DTYPE) * self.d\n",
    "\n",
    "    def offset(self, dx, dy, dz):\n",
    "        \"\"\"\n",
    "        Return the offset in bytes\n",
    "        :param dx: increment in x\n",
    "        :param dy: increment in y\n",
    "        :param dz: increment in z\n",
    "        :return: offset, to be used for as_strided\n",
    "        \"\"\"\n",
    "        return ((dx * self.ny + dy) * self.nz + dz) * self._grid.itemsize\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec9e386f-cd15-446b-8f8a-96b971ecd2cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "class definitions for CavFind\n",
    "\"\"\"\n",
    "import os\n",
    "import re\n",
    "import time\n",
    "from datetime import datetime\n",
    "from logging import getLogger\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "\n",
    "logger = getLogger(__name__)\n",
    "\n",
    "default_settings = {\n",
    "    # PDB interpretation\n",
    "    \"keep_hydrogens\": False,\n",
    "    \"keep_hetatms\": False,\n",
    "    \"keep_waters\": False,\n",
    "    \"resn_water\": (\"HOH\", \"WAT\", \"H2O\"),\n",
    "    \"alternate\": \"A\",  # set to 'all' to include all alternate conformations\n",
    "    # Grid setup\n",
    "    \"grid_spacing\": 0.7,\n",
    "    \"probe_radius\": 1.4,\n",
    "    \"softness\": 0.5,\n",
    "    \"cushion\": 0.0,\n",
    "    \"radii\": \"UA\",\n",
    "    \"radius_factor\": 1.0,\n",
    "    # LigSite\n",
    "    \"ligsite_cutoff\": 5,\n",
    "    \"gap\": 0,\n",
    "    \"original_ligsite\": False,\n",
    "    # should be approx. sqrt(3)*grid_spacing for comparable results\n",
    "    \"max_dist\": 1.22,\n",
    "    # resolution for volume estimation, grid_spacing divided by this number\n",
    "    \"vol_resol\": 3.0,\n",
    "    # 'old_cavfind': False,\n",
    "    \"min_size\": 4,\n",
    "    \"max_size\": 99999,\n",
    "    \"annotate\": True,\n",
    "    # 'min_volume': 4 * 0.7 ** 3,\n",
    "    # 'max_volume': 99999 * 0.7 ** 3,\n",
    "    \"split_files\": False,\n",
    "    # Annotation and Cropping\n",
    "    \"shape_dmax\": 3.0,\n",
    "    \"shape_object\": \"sele\",\n",
    "    \"cp_limit\": 25.0,\n",
    "}\n",
    "\n",
    "tool_tips = {\n",
    "    # PDB interpretation\n",
    "    \"keep_hydrogens\": \"Keep H-atoms for LigSite calculation (default: False).\",\n",
    "    \"keep_hetatms\": \"Keep hetero-components for LigSite calculation (default: False).\",\n",
    "    \"keep_waters\": \"Keep water molecules for LigSite calculation (default: False).\",\n",
    "    \"resn_water\": \"Residue names for 'water'.\",\n",
    "    \"alternate\": \"alternate conformation to be kept (default: 'A'), \"\n",
    "    + \"use 'all' to keep all alternates.\",\n",
    "    # Grid setup\n",
    "    \"grid_spacing\": \"Grid spacing in Angs. (default: 0.7).\",\n",
    "    \"probe_radius\": \"Probe size in Angs. (default: 1.4).\",\n",
    "    \"softness\": \"Soft shell in grid units (default: 0.5).\",\n",
    "    \"cushion\": \"Cushion around the coordinates (default: 0.).\",\n",
    "    \"radii\": \"Atomic radii to use for: 'UA' (united-atoms, default), 'AA' (all-atom).\",\n",
    "    \"radius_factor\": \"Multiplicative factor for atom radii (default: 1.0).\",\n",
    "    # LigSite\n",
    "    \"ligsite_cutoff\": \"Cutoff value for the LigSite algorithm (default: 5).\",\n",
    "    \"gap\": \"Maximum gap for cavity coalescence (default: 0).\",\n",
    "    \"original_ligsite\": \"Find cavities using a distance cutoff.\",\n",
    "    \"max_dist\": \"Maximum distance used in cavity detection (default: 1.22)\",\n",
    "    \"vol_resol\": \"Resolution for volume estimation, higher is more accurate (default: 3.0).\",\n",
    "    # 'old_cavfind': \"Use the old GrowFromSeed algorithm (default: False).\",\n",
    "    \"min_size\": \"Minimum size of a cavity in grid points (default: 4).\",\n",
    "    \"max_size\": \"Maximum size of a cavity in grid points (default: 99999).\",\n",
    "    \"annotate\": \"Annotate cavities with Coulomb potential and hydrobicity. Disable for large structures.\",\n",
    "    # 'min_volume': \"Minimum size of a cavity in Angs.**3.\",\n",
    "    # 'max_volume': \"Maximum size of a cavity in Angs.**3.\",\n",
    "    \"split_files\": \"Generate separate files for each cavity (default: False).\",\n",
    "    # Annotation and Cropping\n",
    "    \"shape_dmax\": \"Maximum distance of a grid point to the shaping object (default: 3.0).\",\n",
    "    \"shape_object\": \"PyMOL object/selection used for cavity shaping (default: sele).\",\n",
    "    \"cp_limit\": \"Limit for displaying the Coulomb potential\",\n",
    "}\n",
    "\n",
    "param_types = {\n",
    "    # types of setting parameters:\n",
    "    # 0 ... boolean\n",
    "    # 1 ... integer\n",
    "    # 2 ... float\n",
    "    # 3 ... str\n",
    "    # 4 ... tuple\n",
    "    # PDB interpretation\n",
    "    \"keep_hydrogens\": 0,\n",
    "    \"keep_hetatms\": 0,\n",
    "    \"keep_waters\": 0,\n",
    "    \"resn_water\": 4,\n",
    "    \"alternate\": 3,\n",
    "    # Grid setup\n",
    "    \"grid_spacing\": 2,\n",
    "    \"probe_radius\": 2,\n",
    "    \"softness\": 2,\n",
    "    \"cushion\": 2,\n",
    "    \"radii\": 3,\n",
    "    \"radius_factor\": 2,\n",
    "    # LigSite\n",
    "    \"ligsite_cutoff\": 1,\n",
    "    \"gap\": 1,\n",
    "    \"original_ligsite\": 0,\n",
    "    \"max_dist\": 2,\n",
    "    \"vol_resol\": 2,\n",
    "    # 'old_cavfind': 0,\n",
    "    \"min_size\": 1,\n",
    "    \"max_size\": 1,\n",
    "    \"annotate\": 0,\n",
    "    # 'min_volume': 2,\n",
    "    # 'max_volume': 2,\n",
    "    \"split_files\": 0,\n",
    "    \"shape_dmax\": 2,\n",
    "    \"shape_object\": 3,\n",
    "    \"cp_limit\": 2,\n",
    "}\n",
    "\n",
    "# global parameters\n",
    "GRID_DTYPE = np.int8  # numerical type of the grid object\n",
    "PROTEIN_FLAG = -100  # flag for grid points occupied by the protein\n",
    "SOFT_FLAG = -99  # flag for grid points in the 'soft shell' around protein atoms\n",
    "\n",
    "\n",
    "class CavFind:\n",
    "    \"\"\"\n",
    "    CavFind object\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, name, obj_name, settings=None):\n",
    "        self.name = name\n",
    "        self.obj_name = obj_name\n",
    "        self.time_of_creation = datetime.now()\n",
    "\n",
    "        if settings is not None:\n",
    "            self.settings = settings.copy()\n",
    "        else:\n",
    "            self.settings = default_settings.copy()\n",
    "\n",
    "        # structure for which cavities should be calculated (PDBStructure)\n",
    "        self.pdb = None\n",
    "\n",
    "        self.coords = None  # atom coordinates as numpy array\n",
    "        self.radii = None  # atom radii as numpy array\n",
    "        self.hp = None  # hydrophobicity parameters as numpy array\n",
    "        self.charges = None  # partial charges\n",
    "\n",
    "        self.grid = None  # grid object for LigSite algorithm\n",
    "        self.cavities = None  # cavities detected in structure\n",
    "\n",
    "    def struct_from_pdb(self, file_obj):\n",
    "        \"\"\"\n",
    "        import a structure, assuming PDB-formatted data\n",
    "        clean the structure according to setting parameters:\n",
    "        'keep_hydrogens', 'keep_hetatms', 'keep_waters', 'alternates'\n",
    "\n",
    "        :param file_obj: PDB data, file, list/tuple of lines, PyMOL pdbstr\n",
    "        \"\"\"\n",
    "        self.pdb = PDBStructure(file_obj)\n",
    "\n",
    "        keep_hydrogens = self.settings[\"keep_hydrogens\"]\n",
    "        keep_hetatms = self.settings[\"keep_hetatms\"]\n",
    "        keep_waters = self.settings[\"keep_waters\"]\n",
    "        alternate = self.settings[\"alternate\"]\n",
    "        resn_water = self.settings[\"resn_water\"]\n",
    "\n",
    "        if (\n",
    "            not keep_hydrogens\n",
    "            or not keep_hetatms\n",
    "            or not keep_waters\n",
    "            or alternate != \"all\"\n",
    "        ):  # remove at least some atoms\n",
    "            keep_atoms = []\n",
    "            for atom in self.pdb.atom:\n",
    "\n",
    "                alt_flag = atom.alternate in (alternate, \" \", \"\") or alternate == \"all\"\n",
    "                if not alt_flag:  # only keep atoms with the correct alternate flag\n",
    "                    continue\n",
    "\n",
    "                is_hydrogen = re.match(\"^[0-9H]\", atom.name.strip()) is not None\n",
    "                is_water = atom.residue in resn_water\n",
    "\n",
    "                if not is_hydrogen and not atom.het:\n",
    "                    # is not a hydrogen, not a HETATM and has the 'correct' alternate flag\n",
    "                    keep_atoms.append(atom)\n",
    "                    continue\n",
    "\n",
    "                if is_water and not is_hydrogen and keep_waters:\n",
    "                    # is a water, not a hydrogen and waters should be kept\n",
    "                    keep_atoms.append(atom)\n",
    "                    continue\n",
    "\n",
    "                if atom.het and not is_hydrogen and not is_water and keep_hetatms:\n",
    "                    # is a HETATM, not a hydrogen and HETATMs should be kept\n",
    "                    keep_atoms.append(atom)\n",
    "                    continue\n",
    "\n",
    "                if is_hydrogen and keep_hydrogens:\n",
    "                    # is a hydrogen and hydrogen should be kept\n",
    "                    keep_atoms.append(atom)\n",
    "\n",
    "            self.pdb.atom = keep_atoms\n",
    "\n",
    "        rad_column = {\"UA\": 0, \"AA\": 1}.get(self.settings[\"radii\"], 0)\n",
    "        self.pdb.assign_rad_hp_chg(rad_column=rad_column)\n",
    "\n",
    "        # Generate numpy arrays for coordinates, radii and hydrophobicity parameters\n",
    "        coords = []\n",
    "        radii = []\n",
    "        hp = []\n",
    "        charges = []\n",
    "        for atom in self.pdb.atom:\n",
    "            coords.append((atom.x, atom.y, atom.z))\n",
    "            radii.append(atom.radius)\n",
    "            hp.append(atom.HP)\n",
    "            charges.append(atom.charge)\n",
    "        self.coords = np.array(coords, dtype=FP_DTYPE)\n",
    "        self.radii = np.array(radii, dtype=FP_DTYPE)\n",
    "        self.hp = np.array(hp, dtype=FP_DTYPE)\n",
    "        self.charges = np.array(charges, dtype=FP_DTYPE)\n",
    "        logger.info(f\"Total charge: {np.sum(self.charges):2f}\")\n",
    "\n",
    "    def run(self, rerun=False, progress_callback=None):\n",
    "        \"\"\"\n",
    "        perform a LigSite calculation and find cavities\n",
    "        if re_run is True, rerun cavity detection on existing grid,\n",
    "        e.g. with different 'ligsite_cutoff', 'gap_parameter' or 'min/max_size'\n",
    "        old cavities are deleted!\n",
    "        \"\"\"\n",
    "        if not rerun:\n",
    "            # perform a complete Ligsite calculation\n",
    "            # setup and mask the grid\n",
    "            self._emit_message(f\"{self.name}:\\nsetup grid\", progress_callback)\n",
    "\n",
    "            # setup and mask the grid\n",
    "            start = time.perf_counter()\n",
    "            self.grid = setup_grid(\n",
    "                self.coords,\n",
    "                self.settings[\"cushion\"],\n",
    "                self.settings[\"grid_spacing\"],\n",
    "                0,\n",
    "                GRID_DTYPE,\n",
    "            )\n",
    "            mask_grid(\n",
    "                self.grid,\n",
    "                self.coords,\n",
    "                self.radii,\n",
    "                self.settings[\"radius_factor\"],\n",
    "                self.settings[\"probe_radius\"],\n",
    "                self.settings[\"softness\"],\n",
    "                PROTEIN_FLAG,\n",
    "                SOFT_FLAG,\n",
    "            )\n",
    "            stop = time.perf_counter()\n",
    "            logger.info(f\"{self.name}: {stop - start:.2f} sec.\")\n",
    "\n",
    "            # run the LigSite algorithm and find cavities\n",
    "            self._emit_message(\n",
    "                f\"{self.name}:\\nanalyse grid\\n({np.prod(self.grid.extent)} points)\",\n",
    "                progress_callback,\n",
    "            )\n",
    "\n",
    "            start = time.perf_counter()\n",
    "            do_ligsite(self.grid, PROTEIN_FLAG)\n",
    "            stop = time.perf_counter()\n",
    "            logger.info(f\"{self.name}: {stop - start:.2f} sec.\")\n",
    "\n",
    "        # detect cavities in the grid (entry point in case of re-run)\n",
    "        super(CavFind, self).__init__\n",
    "        self._emit_message(\n",
    "            f\"{self.name}:\\ndetect cavities\\n(cutoff: {self.settings['ligsite_cutoff']})\",\n",
    "            progress_callback,\n",
    "        )\n",
    "\n",
    "        start = time.perf_counter()\n",
    "        if self.settings[\"original_ligsite\"]:\n",
    "            self.cavities = find_cavities_dist(\n",
    "                self.grid,\n",
    "                self.settings[\"ligsite_cutoff\"],\n",
    "                self.settings[\"max_dist\"],\n",
    "                self.settings[\"min_size\"],\n",
    "                self.settings[\"max_size\"],\n",
    "                self.settings[\"probe_radius\"],\n",
    "                self.settings[\"vol_resol\"],\n",
    "            )\n",
    "        else:\n",
    "            self.cavities = find_cavities(\n",
    "                self.grid,\n",
    "                self.settings[\"ligsite_cutoff\"],\n",
    "                self.settings[\"gap\"],\n",
    "                self.settings[\"min_size\"],\n",
    "                self.settings[\"max_size\"],\n",
    "                self.settings[\"probe_radius\"],\n",
    "                self.settings[\"vol_resol\"],\n",
    "            )\n",
    "        stop = time.perf_counter()\n",
    "        logger.info(f\"{self.name}: {stop - start:.2f} sec.\")\n",
    "\n",
    "        if self.settings[\"annotate\"]:\n",
    "            # annotate cavities\n",
    "            self._emit_message(\n",
    "                f\"{self.name}:\\nannotate\\n({len(self.cavities)} cavities)\",\n",
    "                progress_callback,\n",
    "            )\n",
    "\n",
    "            start = time.perf_counter()\n",
    "            for cav in self.cavities:\n",
    "                cav.annotate(self.coords, self.charges, self.hp)\n",
    "            stop = time.perf_counter()\n",
    "            logger.info(f\"{self.name}: {stop - start:.2f} sec.\")\n",
    "        else:\n",
    "            logger.info(f\"{self.name}: {len(self.cavities)} cavities detected.\")\n",
    "\n",
    "        return self  # necessary for asynchronous execution\n",
    "\n",
    "    @staticmethod\n",
    "    def _emit_message(message, progress_callback):\n",
    "        if progress_callback:\n",
    "            progress_callback.emit(message)\n",
    "        logger.info(message.replace(\"\\n\", \" \"))\n",
    "\n",
    "    def _get_pdb_header(self):\n",
    "        \"\"\"\n",
    "        Generate REMARK lines for PDB-output,\n",
    "        containing the settings used in the calculation\n",
    "        :return: string with REMARK lines\n",
    "        \"\"\"\n",
    "        header = [\n",
    "            \"REMARK\",\n",
    "            f\"REMARK     alternate flag: {self.settings['alternate']}\",\n",
    "        ]\n",
    "\n",
    "        if not self.settings[\"keep_hydrogens\"]:\n",
    "            header.append(\"REMARK                     *** hydrogen atoms excluded\")\n",
    "        if not self.settings[\"keep_waters\"]:\n",
    "            header.append(\"REMARK                     *** water molecules excluded\")\n",
    "        if not self.settings[\"keep_hetatms\"]:\n",
    "            header.append(\"REMARK                     *** hetero atoms excluded\")\n",
    "\n",
    "        header.extend(\n",
    "            [\n",
    "                \"REMARK\",\n",
    "                f\"REMARK       grid spacing:{self.settings['grid_spacing']:8.3f}\",\n",
    "                f\"REMARK            cushion:{self.settings['cushion']:7.2f}\",\n",
    "                f\"REMARK       probe radius:{self.settings['probe_radius']:7.2f}\",\n",
    "                f\"REMARK           softness:{self.settings['softness']:7.2f}\",\n",
    "                f\"REMARK      radius factor:{self.settings['radius_factor']:7.2f}\",\n",
    "                \"REMARK\",\n",
    "                f\"REMARK      cavity cutoff:{self.settings['ligsite_cutoff']:7.2f}\",\n",
    "                f\"REMARK      gap parameter:{self.settings['gap']:7.2f}\",\n",
    "                f\"REMARK   min. cavity size:{self.settings['min_size']:7d} points\",\n",
    "                f\"REMARK   max. cavity size:{self.settings['max_size']:7d} points\",\n",
    "                \"REMARK\",\n",
    "                \"\",\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        return \"\\n\".join(header)\n",
    "\n",
    "    def write_cavities(\n",
    "        self,\n",
    "        filename=\"cav.pdb\",\n",
    "        file_format=\"pdb\",\n",
    "        indices=None,\n",
    "        annotation=None,\n",
    "        header=\"\",\n",
    "    ):\n",
    "        \"\"\"\n",
    "\n",
    "        :param indices:\n",
    "        :param filename:\n",
    "        :param file_format:\n",
    "        :param annotation:\n",
    "        :param header:\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        # build list of annotations\n",
    "        annotation_list = []\n",
    "        if isinstance(annotation, list) or isinstance(annotation, tuple):\n",
    "            annotation_list = list(annotation)\n",
    "        elif isinstance(annotation, str):\n",
    "            annotation_list = [\n",
    "                item.strip() for item in annotation.split(\",\") if item.strip()\n",
    "            ]\n",
    "\n",
    "        if \"LIG\" not in annotation_list:  # always save the LigSite score\n",
    "            annotation_list.append(\"LIG\")\n",
    "\n",
    "        # remove unavailable annotations\n",
    "        annotation_list = [\n",
    "            key for key in annotation_list if key in self.cavities[0].annotations\n",
    "        ]\n",
    "\n",
    "        basename, extension = os.path.splitext(filename)\n",
    "\n",
    "        if file_format == \"pdb\":\n",
    "            if extension != \".pdb\":\n",
    "                basename = filename\n",
    "                extension = \".pdb\"\n",
    "            pdb_header = self._get_pdb_header()\n",
    "\n",
    "            for annotation in annotation_list:\n",
    "                annotation_header = f\"REMARK Annotation: {annotation}\\n\"\n",
    "                # write one file with all selected cavities included\n",
    "                if annotation == \"LIG\":  # naming compatibility with CavFind\n",
    "                    fname = f\"{basename}{extension}\"\n",
    "                else:\n",
    "                    fname = f\"{basename}_{annotation}{extension}\"\n",
    "\n",
    "                with open(fname, \"w\") as file:\n",
    "                    file.write(header)\n",
    "                    file.write(pdb_header)\n",
    "                    file.write(annotation_header)\n",
    "                    for i, cav in enumerate(self.cavities):\n",
    "                        # either write all cavities or those in 'indices'\n",
    "                        if indices is None or i in indices:\n",
    "                            cav.write(file, file_format, annotation)\n",
    "                    file.write(\"END\\n\")\n",
    "                    logger.info(\"Cavities written to: {:s}\".format(fname))\n",
    "\n",
    "                if self.settings[\"split_files\"]:\n",
    "                    # write a separate file for each selected cavity\n",
    "                    for i, cav in enumerate(self.cavities):\n",
    "                        # either write all cavities or those in 'indices'\n",
    "                        if indices is None or i in indices:\n",
    "                            if annotation == \"LIG\":  # naming compatibility with CavFind\n",
    "                                fname = f\"{basename}_{i + 1:04d}{extension}\"\n",
    "                            else:\n",
    "                                fname = (\n",
    "                                    f\"{basename}_{annotation}_{i + 1:04d}{extension}\"\n",
    "                                )\n",
    "\n",
    "                            with open(fname, \"w\") as file:\n",
    "                                file.write(header)\n",
    "                                file.write(pdb_header)\n",
    "                                file.write(annotation_header)\n",
    "                                cav.write(file, file_format, annotation)\n",
    "                                file.write(\"END\\n\")\n",
    "                                logger.info(f\"Cavity #{i + 1:d} written to: {fname}\")\n",
    "\n",
    "        # separate file for each cavity with all selected annotations included\n",
    "        elif file_format == \"csv\":\n",
    "            if extension != \".csv\":\n",
    "                basename = filename\n",
    "                extension = \".csv\"\n",
    "\n",
    "            content = \"# columns: x, y, z, \" + \", \".join(annotation_list) + \"\\n\"\n",
    "\n",
    "            for i, cav in enumerate(self.cavities):\n",
    "                # either write all cavities or those in 'indices'\n",
    "                if indices is None or i in indices:\n",
    "                    fname = f\"{basename}_{i + 1:04d}{extension}\"\n",
    "                    with open(fname, \"w\") as file:\n",
    "                        file.write(header)\n",
    "                        file.write(f\"# number of grid points:{cav.size:6d}\\n\")\n",
    "                        file.write(content)\n",
    "                        cav.write(file, \"csv\", annotation_list)\n",
    "                        logger.info(f\"Cavity #{i + 1} written to: {fname}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d562b705-45f2-4ab9-a175-46ef93d68c06",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = CavFind('name', 'obj_name')\n",
    "with open('path/to/file.pdb', 'r') as file:\n",
    "    pdb_string = file.read()\n",
    "a.struct_from_pdb(pdb_string)\n",
    "a.run()\n",
    "a.write_cavities()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
